// @injectivelabs/sdk-ts/dist/cjs/client/chain/transformers/ChainGrpcGovTransformer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcGovTransformer = void 0;
const utils_1 = require("../../../utils");
const pagination_1 = require("../../../utils/pagination");
const utils_2 = require("../../../utils");
/**
 * @category Chain Grpc Transformer
 */
class ChainGrpcGovTransformer {
    static moduleParamsResponseToModuleParams(response) {
        var _a, _b;
        const depositParams = response.depositParams;
        const votingParams = response.votingParams;
        const tallyParams = response.tallyParams;
        return {
            depositParams: {
                minDepositList: depositParams === null || depositParams === void 0 ? void 0 : depositParams.minDeposit,
                maxDepositPeriod: parseInt(((_a = depositParams === null || depositParams === void 0 ? void 0 : depositParams.maxDepositPeriod) === null || _a === void 0 ? void 0 : _a.seconds) || '0', 10),
            },
            votingParams: {
                votingPeriod: parseInt(((_b = votingParams.votingPeriod) === null || _b === void 0 ? void 0 : _b.seconds) || '0'),
            },
            tallyParams: {
                quorum: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.quorum)).toFixed(),
                threshold: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.threshold)).toFixed(),
                vetoThreshold: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.vetoThreshold)).toFixed(),
            },
        };
    }
    static moduleParamsResponseToModuleParamsByType({ depositParams, votingParams, tallyParams, }) {
        var _a, _b;
        return {
            depositParams: {
                minDepositList: depositParams === null || depositParams === void 0 ? void 0 : depositParams.minDeposit,
                maxDepositPeriod: parseInt(((_a = depositParams === null || depositParams === void 0 ? void 0 : depositParams.maxDepositPeriod) === null || _a === void 0 ? void 0 : _a.seconds) || '0', 10),
            },
            votingParams: {
                votingPeriod: parseInt(((_b = votingParams.votingPeriod) === null || _b === void 0 ? void 0 : _b.seconds) || '0'),
            },
            tallyParams: {
                quorum: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.quorum)).toFixed(),
                threshold: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.threshold)).toFixed(),
                vetoThreshold: (0, utils_2.cosmosSdkDecToBigNumber)((0, utils_1.uint8ArrayToString)(tallyParams.vetoThreshold)).toFixed(),
            },
        };
    }
    static proposalResponseToProposal(response) {
        const proposal = response.proposal;
        return ChainGrpcGovTransformer.grpcProposalToProposal(proposal);
    }
    static proposalsResponseToProposals(response) {
        const proposals = response.proposals.map((p) => ChainGrpcGovTransformer.grpcProposalToProposal(p));
        const pagination = response.pagination;
        return {
            proposals: proposals,
            pagination: (0, pagination_1.grpcPaginationToPagination)(pagination),
        };
    }
    static depositsResponseToDeposits(response) {
        const pagination = response.pagination;
        const deposits = response.deposits.map((deposit) => {
            return {
                depositor: deposit.depositor,
                amounts: deposit.amount.map((coin) => ({
                    denom: coin.denom,
                    amount: (0, utils_2.cosmosSdkDecToBigNumber)(coin.amount).toFixed(),
                })),
            };
        });
        return {
            deposits: deposits,
            pagination: (0, pagination_1.grpcPaginationToPagination)(pagination),
        };
    }
    static votesResponseToVotes(response) {
        const pagination = response.pagination;
        const votes = response.votes.map((v) => {
            return {
                proposalId: v.proposalId,
                voter: v.voter,
                option: v.option,
            };
        });
        return {
            votes: votes.map((v) => (Object.assign(Object.assign({}, v), { proposalId: parseInt(v.proposalId, 10) }))),
            pagination: (0, pagination_1.grpcPaginationToPagination)(pagination),
        };
    }
    static tallyResultResponseToTallyResult(response) {
        const result = response.tally;
        return ChainGrpcGovTransformer.grpcTallyResultToTallyResult(result);
    }
    static grpcTallyResultToTallyResult(result) {
        return {
            yes: result ? result.yes : '0',
            abstain: result ? result.abstain : '0',
            no: result ? result.no : '0',
            noWithVeto: result ? result.noWithVeto : '0',
        };
    }
    static grpcProposalToProposal(proposal) {
        const finalTallyResult = proposal.finalTallyResult;
        const content = proposal.content;
        return {
            proposalId: parseInt(proposal.proposalId, 10),
            content: {
                type: content.typeUrl,
                value: content.value,
            },
            type: content.typeUrl,
            submitTime: Math.floor(proposal.submitTime.getTime() / 1000),
            status: proposal.status,
            finalTallyResult: ChainGrpcGovTransformer.grpcTallyResultToTallyResult(finalTallyResult),
            depositEndTime: Math.floor(proposal.depositEndTime.getTime() / 1000),
            totalDeposits: proposal.totalDeposit.map((coin) => ({
                denom: coin.denom,
                amount: (0, utils_2.cosmosSdkDecToBigNumber)(coin.amount).toFixed(),
            })),
            votingStartTime: Math.floor(proposal.votingStartTime.getTime() / 1000),
            votingEndTime: Math.floor(proposal.votingEndTime.getTime() / 1000),
        };
    }
}
exports.ChainGrpcGovTransformer = ChainGrpcGovTransformer;
