// @certusone/wormhole-sdk/lib/esm/ethers-contracts/factories/State.sol/Events__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { utils, Contract, ContractFactory } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint32",
                name: "oldGuardianIndex",
                type: "uint32",
            },
            {
                indexed: false,
                internalType: "uint32",
                name: "newGuardianIndex",
                type: "uint32",
            },
        ],
        name: "LogGuardianSetChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "emitter_address",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint32",
                name: "nonce",
                type: "uint32",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
        ],
        name: "LogMessagePublished",
        type: "event",
    },
];
const _bytecode = "0x60808060405234601357603a908160198239f35b600080fdfe600080fdfea26469706673582212205e1d7f0c8904733b3c86668b39c62ef6e460cccbb522f631eebb1af6fcaa11f164736f6c63430008130033";
const isSuperArgs = (xs) => xs.length > 1;
export class Events__factory extends ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
Events__factory.bytecode = _bytecode;
Events__factory.abi = _abi;
