// @injectivelabs/sdk-ts/dist/cjs/core/modules/wasm/exec-args/ExecArgUpdateAMMVaultConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ExecArgBase_1 = require("../ExecArgBase");
/**
 * @category Contract Exec Arguments
 */
class ExecArgUpdateAMMVaultConfig extends ExecArgBase_1.ExecArgBase {
    static fromJSON(params) {
        return new ExecArgUpdateAMMVaultConfig(params);
    }
    toData() {
        const { params } = this;
        return {
            owner: params.owner,
            market_id: params.marketId,
            master_address: params.masterAddress,
            order_density: params.orderDensity,
            price_tick_size: params.priceTickSize,
            max_invariant_sensitivity: params.maxInvariantSensitivity,
            base_decimals: params.baseDecimals,
            quote_decimals: params.quoteDecimals,
            notional_value_cap: params.notionalValueCap,
        };
    }
    toExecData() {
        return (0, ExecArgBase_1.dataToExecData)('update_vault_config', this.toData());
    }
}
exports.default = ExecArgUpdateAMMVaultConfig;
