// @injectivelabs/sdk-ts/dist/cjs/core/modules/wasm/exec-args/ExecArgUpdateDerivativeVaultConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ExecArgBase_1 = require("../ExecArgBase");
/**
 * @category Contract Exec Arguments
 */
class ExecArgUpdateDerivativeVaultConfig extends ExecArgBase_1.ExecArgBase {
    static fromJSON(params) {
        return new ExecArgUpdateDerivativeVaultConfig(params);
    }
    toData() {
        const { params } = this;
        return {
            market_id: params.marketId,
            leverage: params.leverage,
            order_density: params.orderDensity,
            signed_min_head_to_fair_price_deviation_ratio: params.signedMinHeadToFairPriceDeviationRatio,
            signed_min_head_to_tob_deviation_ratio: params.signedMinHeadToTobDeviationRatio,
            reservation_price_sensitivity_ratio: params.reservationPriceSensitivityRatio,
            reservation_spread_sensitivity_ratio: params.reservationSpreadSensitivityRatio,
            max_active_capital_utilization_ratio: params.maxActiveCapitalUtilizationRatio,
            head_change_tolerance_ratio: params.headChangeToleranceRatio,
            min_head_to_tail_deviation_ratio: params.minHeadToTailDeviationRatio,
            min_proximity_to_liquidation: params.minProximityToLiquidation,
            post_reduction_perc_of_max_position: params.postReductionPercOfMaxPosition,
            oracle_volatility_group_sec: params.oracleVolatilityGroupSec,
            min_oracle_volatility_sample_size: params.minOracleVolatilitySampleSize,
            emergency_oracle_volatility_sample_size: params.emergencyOracleVolatilitySampleSize,
            trade_volatility_group_sec: params.tradeVolatilityGroupSec,
            min_trade_volatility_sample_size: params.minTradeVolatilitySampleSize,
            default_mid_price_volatility_ratio: params.defaultMidPriceVolatilityRatio,
            min_volatility_ratio: params.minVolatilityRatio,
            last_valid_mark_price: params.lastValidMarkPrice,
            allowed_redemption_types: params.allowedRedemptionTypes,
            position_pnl_penalty: params.positionPnlPenalty,
            master_address: params.masterAddress,
            owner: params.owner,
            quote_decimals: params.quoteDecimals,
            notional_value_cap: params.notionalValueCap,
        };
    }
    toExecData() {
        return (0, ExecArgBase_1.dataToExecData)('update_vault_config', this.toData());
    }
}
exports.default = ExecArgUpdateDerivativeVaultConfig;
