// @injectivelabs/sdk-ts/dist/cjs/core/modules/wasm/exec-args/ExecArgUpdateSpotVaultConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ExecArgBase_1 = require("../ExecArgBase");
/**
 * @category Contract Exec Arguments
 */
class ExecArgUpdateSpotVaultConfig extends ExecArgBase_1.ExecArgBase {
    static fromJSON(params) {
        return new ExecArgUpdateSpotVaultConfig(params);
    }
    toData() {
        const { params } = this;
        return {
            market_id: params.marketId,
            order_density: params.orderDensity,
            reservation_price_sensitivity_ratio: params.reservationPriceSensitivityRatio,
            reservation_spread_sensitivity_ratio: params.reservationSpreadSensitivityRatio,
            max_active_capital_utilization_ratio: params.maxActiveCapitalUtilizationRatio,
            head_change_tolerance_ratio: params.headChangeToleranceRatio,
            reservation_price_tail_deviation_ratio: params.reservationPriceTailDeviationRatio,
            min_head_to_tail_deviation_ratio: params.minHeadToTailDeviationRatio,
            min_volatility_ratio: params.minVolatilityRatio,
            signed_min_head_to_fair_price_deviation_ratio: params.signedMinHeadToFairPriceDeviationRatio,
            signed_min_head_to_tob_deviation_ratio: params.signedMinHeadToTobDeviationRatio,
            target_base_weight: params.targetBaseWeight,
            trade_volatility_group_sec: params.tradeVolatilityGroupSec,
            min_trade_volatility_sample_size: params.minTradeVolatilitySampleSize,
            default_mid_price_volatility_ratio: params.defaultMidPriceVolatilityRatio,
            oracle_type: params.oracleType,
            allowed_redemption_types: params.allowedRedemptionTypes,
            base_decimals: params.baseDecimals,
            quote_decimals: params.quoteDecimals,
            base_oracle_symbol: params.baseOracleSymbol,
            quote_oracle_symbol: params.quoteOracleSymbol,
            master_address: params.masterAddress,
            owner: params.owner,
            notional_value_cap: params.notionalValueCap,
        };
    }
    toExecData() {
        return (0, ExecArgBase_1.dataToExecData)('update_vault_config', this.toData());
    }
}
exports.default = ExecArgUpdateSpotVaultConfig;
