// @injectivelabs/sdk-ts/dist/cjs/core/modules/wasm/exec-priv-args/ExecPrivilegedArgVaultRedeem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ExecPrivilegedArgBase_1 = require("../ExecPrivilegedArgBase");
/**
 * @category Contract Exec Arguments
 */
class ExecPrivilegedArgVaultRedeem extends ExecPrivilegedArgBase_1.ExecPrivilegedArgBase {
    static fromJSON(params) {
        return new ExecPrivilegedArgVaultRedeem(params);
    }
    toData() {
        const { params } = this;
        return {
            slippage: params.slippage,
            redemption_type: params.redemptionType,
            redemption_ratio: params.redemptionRatio,
            vault_subaccount_id: params.vaultSubaccountId,
            redeemer_subaccount_id: params.redeemerSubaccountId,
        };
    }
    toExecData() {
        const { params } = this;
        return (0, ExecPrivilegedArgBase_1.dataToExecData)(this.toData(), {
            origin: params.origin,
            name: 'VaultRedeem',
            action: 'Redeem',
        });
    }
}
exports.default = ExecPrivilegedArgVaultRedeem;
