// @across-protocol/sdk-v2/dist/esm/contracts/utils.js
import assert from "assert";
import { BigNumber } from "ethers";
import { isDefined } from "../utils";
/**
 * Utility for maintaining records of a token balance.
 */
export function Balances(balances) {
    if (balances === void 0) { balances = {}; }
    function create(id, amount) {
        if (amount === void 0) { amount = "0"; }
        assert(!has(id), "balance already exists");
        return set(id, amount);
    }
    function has(id) {
        return isDefined(balances[id]);
    }
    function set(id, amount) {
        balances[id] = amount;
        return amount;
    }
    function add(id, amount) {
        return set(id, BigNumber.from(amount).add(getOrCreate(id)).toString());
    }
    function sub(id, amount) {
        return set(id, BigNumber.from(getOrCreate(id)).sub(amount).toString());
    }
    function get(id) {
        assert(has(id), "balance does not exist");
        return balances[id];
    }
    function getOrCreate(id) {
        if (has(id))
            return get(id);
        return create(id);
    }
    return { create: create, add: add, sub: sub, get: get, balances: balances, set: set, has: has, getOrCreate: getOrCreate };
}
