// @uma/contracts-frontend/dist/typechain/core/ethers/factories/ConfigStoreInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "updateAndGetCurrentConfig",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "timelockLiveness",
                        type: "uint256",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256",
                                name: "rawValue",
                                type: "uint256",
                            },
                        ],
                        internalType: "struct FixedPoint.Unsigned",
                        name: "rewardRatePerSecond",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "uint256",
                                name: "rawValue",
                                type: "uint256",
                            },
                        ],
                        internalType: "struct FixedPoint.Unsigned",
                        name: "proposerBondPercentage",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "int256",
                                name: "rawValue",
                                type: "int256",
                            },
                        ],
                        internalType: "struct FixedPoint.Signed",
                        name: "maxFundingRate",
                        type: "tuple",
                    },
                    {
                        components: [
                            {
                                internalType: "int256",
                                name: "rawValue",
                                type: "int256",
                            },
                        ],
                        internalType: "struct FixedPoint.Signed",
                        name: "minFundingRate",
                        type: "tuple",
                    },
                    {
                        internalType: "uint256",
                        name: "proposalTimePastLimit",
                        type: "uint256",
                    },
                ],
                internalType: "struct ConfigStoreInterface.ConfigSettings",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class ConfigStoreInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
ConfigStoreInterface__factory.abi = _abi;
