// @uma/contracts-frontend/dist/typechain/core/ethers/factories/FinderInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "interfaceName",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "implementationAddress",
                type: "address",
            },
        ],
        name: "changeImplementationAddress",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "interfaceName",
                type: "bytes32",
            },
        ],
        name: "getImplementationAddress",
        outputs: [
            {
                internalType: "address",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class FinderInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
FinderInterface__factory.abi = _abi;
