// @uma/contracts-frontend/dist/typechain/core/ethers/factories/IGenericHandler__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "resourceID",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "contractAddress",
                type: "address",
            },
            {
                internalType: "bytes4",
                name: "depositFunctionSig",
                type: "bytes4",
            },
            {
                internalType: "bytes4",
                name: "executeFunctionSig",
                type: "bytes4",
            },
        ],
        name: "setResource",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class IGenericHandler__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
IGenericHandler__factory.abi = _abi;
