// @uma/contracts-frontend/dist/typechain/core/ethers/factories/MinimumVotingAncillaryInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "voterAddress",
                type: "address",
            },
            {
                internalType: "uint256",
                name: "roundId",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "bytes32",
                        name: "identifier",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint256",
                        name: "time",
                        type: "uint256",
                    },
                    {
                        internalType: "bytes",
                        name: "ancillaryData",
                        type: "bytes",
                    },
                ],
                internalType: "struct MinimumVotingAncillaryInterface.PendingRequestAncillary[]",
                name: "toRetrieve",
                type: "tuple[]",
            },
        ],
        name: "retrieveRewards",
        outputs: [
            {
                components: [
                    {
                        internalType: "uint256",
                        name: "rawValue",
                        type: "uint256",
                    },
                ],
                internalType: "struct MinimumVotingAncillaryInterface.Unsigned",
                name: "",
                type: "tuple",
            },
        ],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class MinimumVotingAncillaryInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
MinimumVotingAncillaryInterface__factory.abi = _abi;
