// @uma/contracts-frontend/dist/typechain/core/ethers/factories/OptimisticOracleV3CallbackRecipientInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "assertionId",
                type: "bytes32",
            },
        ],
        name: "assertionDisputedCallback",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "assertionId",
                type: "bytes32",
            },
            {
                internalType: "bool",
                name: "assertedTruthfully",
                type: "bool",
            },
        ],
        name: "assertionResolvedCallback",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class OptimisticOracleV3CallbackRecipientInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
OptimisticOracleV3CallbackRecipientInterface__factory.abi = _abi;
