// @wormhole-foundation/sdk-base/dist/cjs/constants/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.guardians = exports.contracts = exports.circle = exports.nativeChainIds = exports.rpc = exports.explorer = exports.decimals = exports.finality = exports.network = exports.chain = exports.platform = exports.tokens = exports.platformToAddressFormat = exports.chainToPlatform = exports.platformToChains = exports.isPlatform = exports.chainIdToChain = exports.chainToChainId = exports.chains = exports.toChain = exports.toChainId = exports.isChain = void 0;
var chains_1 = require("./chains");
Object.defineProperty(exports, "isChain", { enumerable: true, get: function () { return chains_1.isChain; } });
Object.defineProperty(exports, "toChainId", { enumerable: true, get: function () { return chains_1.toChainId; } });
Object.defineProperty(exports, "toChain", { enumerable: true, get: function () { return chains_1.toChain; } });
Object.defineProperty(exports, "chains", { enumerable: true, get: function () { return chains_1.chains; } });
Object.defineProperty(exports, "chainToChainId", { enumerable: true, get: function () { return chains_1.chainToChainId; } });
Object.defineProperty(exports, "chainIdToChain", { enumerable: true, get: function () { return chains_1.chainIdToChain; } });
var platforms_1 = require("./platforms");
Object.defineProperty(exports, "isPlatform", { enumerable: true, get: function () { return platforms_1.isPlatform; } });
Object.defineProperty(exports, "platformToChains", { enumerable: true, get: function () { return platforms_1.platformToChains; } });
Object.defineProperty(exports, "chainToPlatform", { enumerable: true, get: function () { return platforms_1.chainToPlatform; } });
Object.defineProperty(exports, "platformToAddressFormat", { enumerable: true, get: function () { return platforms_1.platformToAddressFormat; } });
exports.tokens = __importStar(require("./tokens"));
exports.platform = __importStar(require("./platforms"));
exports.chain = __importStar(require("./chains"));
exports.network = __importStar(require("./networks"));
exports.finality = __importStar(require("./finality"));
exports.decimals = __importStar(require("./decimals"));
exports.explorer = __importStar(require("./explorer"));
exports.rpc = __importStar(require("./rpc"));
exports.nativeChainIds = __importStar(require("./nativeChainIds"));
exports.circle = __importStar(require("./circle"));
exports.contracts = __importStar(require("./contracts"));
exports.guardians = __importStar(require("./guardians"));
