// @certusone/wormhole-sdk/lib/esm/ethers-contracts/factories/WormholeRelayerGovernance.t.sol/Brick__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { utils, Contract, ContractFactory } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "checkAndExecuteUpgradeMigration",
        outputs: [],
        stateMutability: "view",
        type: "function",
    },
];
const _bytecode = "0x60808060405234601357606b908160198239f35b600080fdfe6004361015600c57600080fd5b6000803560e01c632c75470f14602157600080fd5b346032578060031936011260325780f35b80fdfea26469706673582212203dbf06b9c4ea3c944f6459dcad64195f4d25987cff72b6fd25557326823e6b3464736f6c63430008130033";
const isSuperArgs = (xs) => xs.length > 1;
export class Brick__factory extends ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
Brick__factory.bytecode = _bytecode;
Brick__factory.abi = _abi;
