// @certusone/wormhole-sdk/lib/cjs/ethers-contracts/factories/StdStorage.sol/StdStorageSafe__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StdStorageSafe__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "who",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes4",
                name: "fsig",
                type: "bytes4",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "keysHash",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "slot",
                type: "uint256",
            },
        ],
        name: "SlotFound",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "who",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "slot",
                type: "uint256",
            },
        ],
        name: "WARNING_UninitedSlot",
        type: "event",
    },
];
const _bytecode = "0x60808060405234601757603a9081601d823930815050f35b600080fdfe600080fdfea2646970667358221220f35dabdae307840da873431b0101de65906178c6bec9fd28a6c1ea04e23d543e64736f6c63430008130033";
const isSuperArgs = (xs) => xs.length > 1;
class StdStorageSafe__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.StdStorageSafe__factory = StdStorageSafe__factory;
StdStorageSafe__factory.bytecode = _bytecode;
StdStorageSafe__factory.abi = _abi;
