// @wormhole-foundation/sdk-base/dist/cjs/utils/layout/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.customizableBytes = exports.isLayoutItem = exports.isLayout = exports.layoutDiscriminator = exports.addFixedValues = exports.dynamicItemsOfLayout = exports.fixedItemsOfLayout = exports.deserializeLayout = exports.serializeLayout = void 0;
var serialize_1 = require("./serialize");
Object.defineProperty(exports, "serializeLayout", { enumerable: true, get: function () { return serialize_1.serializeLayout; } });
var deserialize_1 = require("./deserialize");
Object.defineProperty(exports, "deserializeLayout", { enumerable: true, get: function () { return deserialize_1.deserializeLayout; } });
var fixedDynamic_1 = require("./fixedDynamic");
Object.defineProperty(exports, "fixedItemsOfLayout", { enumerable: true, get: function () { return fixedDynamic_1.fixedItemsOfLayout; } });
Object.defineProperty(exports, "dynamicItemsOfLayout", { enumerable: true, get: function () { return fixedDynamic_1.dynamicItemsOfLayout; } });
Object.defineProperty(exports, "addFixedValues", { enumerable: true, get: function () { return fixedDynamic_1.addFixedValues; } });
var discriminate_1 = require("./discriminate");
Object.defineProperty(exports, "layoutDiscriminator", { enumerable: true, get: function () { return discriminate_1.layoutDiscriminator; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "isLayout", { enumerable: true, get: function () { return utils_1.isLayout; } });
Object.defineProperty(exports, "isLayoutItem", { enumerable: true, get: function () { return utils_1.isLayoutItem; } });
Object.defineProperty(exports, "customizableBytes", { enumerable: true, get: function () { return utils_1.customizableBytes; } });
__exportStar(require("./size"), exports);
__exportStar(require("./items"), exports);
