// @across-protocol/sdk-v2/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMPTY_MESSAGE = exports.DEFAULT_SIMULATED_RELAYER_ADDRESS_TEST = exports.DEFAULT_SIMULATED_RELAYER_ADDRESS = exports.UBA_BOUNDS_RANGE_MIN = exports.UBA_BOUNDS_RANGE_MAX = exports.DEFAULT_CACHING_SAFE_LAG = exports.DEFAULT_CACHING_TTL = exports.DEFAULT_BLOCKCHAIN_EXPLORER_DOMAIN = exports.PUBLIC_NETWORKS = exports.TESTNET_CHAIN_IDS = exports.PRODUCTION_CHAIN_IDS = exports.PROTOCOL_DEFAULT_CHAIN_ID_INDICES = exports.ARWEAVE_TAG_APP_NAME = exports.TRANSFER_THRESHOLD_MAX_CONFIG_STORE_VERSION = exports.HUBPOOL_CHAIN_ID = exports.SECONDS_PER_YEAR = exports.MAX_SAFE_ALLOWANCE = exports.ZERO_ADDRESS = exports.TOKEN_SYMBOLS_MAP = exports.CHAIN_IDs = void 0;
var tslib_1 = require("tslib");
var acxConstants = tslib_1.__importStar(require("@across-protocol/constants-v2"));
var ethers_1 = require("ethers");
var constants_v2_1 = require("@across-protocol/constants-v2");
Object.defineProperty(exports, "CHAIN_IDs", { enumerable: true, get: function () { return constants_v2_1.CHAIN_IDs; } });
Object.defineProperty(exports, "TOKEN_SYMBOLS_MAP", { enumerable: true, get: function () { return constants_v2_1.TOKEN_SYMBOLS_MAP; } });
exports.ZERO_ADDRESS = ethers_1.constants.AddressZero;
exports.MAX_SAFE_ALLOWANCE = "79228162514264337593543950335";
exports.SECONDS_PER_YEAR = 31557600;
exports.HUBPOOL_CHAIN_ID = 1;
exports.TRANSFER_THRESHOLD_MAX_CONFIG_STORE_VERSION = 1;
exports.ARWEAVE_TAG_APP_NAME = "across-protocol";
exports.PROTOCOL_DEFAULT_CHAIN_ID_INDICES = [1, 10, 137, 288, 42161];
exports.PRODUCTION_CHAIN_IDS = [
    acxConstants.MAINNET_CHAIN_IDs.MAINNET,
    acxConstants.MAINNET_CHAIN_IDs.OPTIMISM,
    acxConstants.MAINNET_CHAIN_IDs.POLYGON,
    acxConstants.MAINNET_CHAIN_IDs.ZK_SYNC,
    acxConstants.MAINNET_CHAIN_IDs.BASE,
    acxConstants.MAINNET_CHAIN_IDs.ARBITRUM,
];
exports.TESTNET_CHAIN_IDS = [
    acxConstants.TESTNET_CHAIN_IDs.GOERLI,
    acxConstants.TESTNET_CHAIN_IDs.ZK_SYNC_GOERLI,
    acxConstants.TESTNET_CHAIN_IDs.ZK_SYNC_SEPOLIA,
    acxConstants.TESTNET_CHAIN_IDs.OPTIMISM_GOERLI,
    acxConstants.TESTNET_CHAIN_IDs.MUMBAI,
    acxConstants.TESTNET_CHAIN_IDs.POLYGON_AMOY,
    acxConstants.TESTNET_CHAIN_IDs.BASE_GOERLI,
    acxConstants.TESTNET_CHAIN_IDs.BASE_SEPOLIA,
    acxConstants.TESTNET_CHAIN_IDs.ARBITRUM_GOERLI,
    acxConstants.TESTNET_CHAIN_IDs.ARBITRUM_SEPOLIA,
    acxConstants.TESTNET_CHAIN_IDs.SEPOLIA,
    acxConstants.TESTNET_CHAIN_IDs.OPTIMISM_SEPOLIA,
];
exports.PUBLIC_NETWORKS = {
    1: {
        name: "mainnet",
        etherscan: "https://etherscan.io",
    },
    5: { name: "goerli", etherscan: "https://goerli.etherscan.io" },
    10: { name: "optimism", etherscan: "https://optimistic.etherscan.io" },
    137: {
        name: "polygon-matic",
        etherscan: "https://polygonscan.com",
    },
    324: { name: "zksync", etherscan: "https://explorer.zksync.io" },
    8453: { name: "base", etherscan: "https://basescan.org" },
    42161: { name: "arbitrum", etherscan: "https://arbiscan.io" },
    43114: { name: "avalanche", etherscan: "https://snowtrace.io" },
    80002: { name: "polygon-amoy", etherscan: "https://www.oklink.com/amoy" },
    84531: { name: "base-goerli", etherscan: "https://basescan.org" },
    84532: { name: "base-sepolia", etherscan: "https://sepolia.basescan.org" },
    421613: { name: "arbitrum-goerli", etherscan: "https://goerli.arbiscan.io" },
    421614: { name: "arbitrum-sepolia", etherscan: "https://sepolia.arbiscan.io" },
    534351: { name: "scroll-sepolia", etherscan: "https://sepolia.scrollscan.com" },
    534352: { name: "scroll", etherscan: "https://scrollscan.com" },
    11155111: { name: "sepolia", etherscan: "https://sepolia.etherscan.io" },
    11155420: { name: "optimism-sepolia", etherscan: "https://sepolia-optimistic.etherscan.io" },
};
exports.DEFAULT_BLOCKCHAIN_EXPLORER_DOMAIN = "https://etherscan.io";
exports.DEFAULT_CACHING_TTL = 60 * 60 * 24 * 7 * 2;
exports.DEFAULT_CACHING_SAFE_LAG = 60 * 60;
exports.UBA_BOUNDS_RANGE_MAX = ethers_1.BigNumber.from(String(Number.MAX_SAFE_INTEGER)).mul(ethers_1.utils.parseEther("1.0"));
exports.UBA_BOUNDS_RANGE_MIN = exports.UBA_BOUNDS_RANGE_MAX.mul(-1);
exports.DEFAULT_SIMULATED_RELAYER_ADDRESS = "0x428AB2BA90Eba0a4Be7aF34C9Ac451ab061AC010";
exports.DEFAULT_SIMULATED_RELAYER_ADDRESS_TEST = "0x9A8f92a830A5cB89a3816e3D267CB7791c16b04D";
exports.EMPTY_MESSAGE = "0x";
