// @injectivelabs/sdk-ts/dist/esm/core/utils/Denom/DenomClientSync.js
import { TokenFactory, TokenMetaUtilsFactory, } from '@injectivelabs/token-metadata';
import { Network } from '@injectivelabs/networks';
/**
 * This client can be used to fetch token
 * denoms in a fully sync way (without API calls)
 *
 * @category Utility Classes
 */
export class DenomClientSync {
    tokenFactory;
    tokenMetaUtils;
    constructor(network = Network.Mainnet) {
        this.tokenFactory = TokenFactory.make(network);
        this.tokenMetaUtils = TokenMetaUtilsFactory.make(network);
    }
    getDenomTokenInfo(denom) {
        return this.tokenFactory.toTokenInfo(denom);
    }
    getDenomToken(denom) {
        return this.tokenFactory.toToken(denom);
    }
    getTokenMetaDataBySymbol(symbol) {
        return this.tokenMetaUtils.getMetaBySymbol(symbol);
    }
    getTokenMetaDataByAddress(address) {
        return this.tokenMetaUtils.getMetaByAddress(address);
    }
    getTokenMetaDataByName(name) {
        return this.tokenMetaUtils.getMetaByName(name);
    }
    getCoinGeckoId(denom) {
        return this.tokenMetaUtils.getCoinGeckoIdFromSymbol(denom);
    }
}
