// @injectivelabs/networks/dist/cjs/old-endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oldEndpointsLocal = exports.oldEndpointsDevnet2 = exports.oldEndpointsDevnet1 = exports.oldEndpointsDevnet = exports.oldEndpointsTestnet = exports.oldEndpointsTestnetOld = exports.oldEndpointsTestnetK8s = exports.oldEndpointsPublic = exports.oldEndpointsStaging = exports.oldEndpointsMainnet = exports.oldEndpointsMainnetLB = exports.oldEndpointsMainnetK8s = void 0;
exports.oldEndpointsMainnetK8s = {
    indexerApi: 'https://k8s.mainnet.exchange.grpc-web.injective.network',
    sentryGrpcApi: 'https://k8s.mainnet.chain.grpc-web.injective.network',
    tendermintApi: 'https://k8s.mainnet.tm.injective.network',
    sentryHttpApi: 'https://k8s.mainnet.lcd.injective.network',
};
exports.oldEndpointsMainnetLB = {
    indexerApi: 'https://k8s.mainnet.exchange.grpc-web.injective.network',
    sentryGrpcApi: 'https://k8s.mainnet.chain.grpc-web.injective.network',
    tendermintApi: 'https://k8s.mainnet.tm.injective.network',
    sentryHttpApi: 'https://k8s.mainnet.lcd.injective.network',
};
exports.oldEndpointsMainnet = {
    indexerApi: 'https://api.injective.network',
    sentryGrpcApi: 'https://grpc.injective.network',
    tendermintApi: 'https://tm.injective.network',
    sentryHttpApi: 'https://lcd.injective.network',
};
exports.oldEndpointsStaging = {
    indexerApi: 'https://staging.api.injective.network',
    sentryGrpcApi: 'https://staging.grpc.injective.network',
    tendermintApi: 'https://staging.tm.injective.network',
    sentryHttpApi: 'https://staging.lcd.injective.network',
};
exports.oldEndpointsPublic = {
    indexerApi: 'https://public.api.injective.network',
    sentryGrpcApi: 'https://public.grpc.injective.network',
    tendermintApi: 'https://tm.injective.network',
    sentryHttpApi: 'https://public.lcd.injective.network',
};
exports.oldEndpointsTestnetK8s = {
    indexerApi: 'https://k8s.testnet.exchange.grpc-web.injective.network',
    sentryGrpcApi: 'https://k8s.testnet.chain.grpc-web.injective.network',
    tendermintApi: 'https://k8s.testnet.tm.injective.network',
    sentryHttpApi: 'https://k8s.testnet.lcd.injective.network',
};
exports.oldEndpointsTestnetOld = {
    indexerApi: 'https://testnet.api.injective.dev',
    sentryGrpcApi: 'https://testnet.grpc.injective.dev',
    tendermintApi: 'https://testnet.tm.injective.dev',
    sentryHttpApi: 'https://testnet.lcd.injective.dev',
};
exports.oldEndpointsTestnet = Object.assign({}, exports.oldEndpointsTestnetK8s);
exports.oldEndpointsDevnet = {
    indexerApi: 'https://devnet.api.injective.dev',
    sentryGrpcApi: 'https://devnet.grpc.injective.dev',
    tendermintApi: 'https://devnet.tm.injective.dev',
    sentryHttpApi: 'https://devnet.lcd.injective.dev',
};
exports.oldEndpointsDevnet1 = {
    indexerApi: 'https://devnet-1.api.injective.dev',
    sentryGrpcApi: 'https://devnet-1.grpc.injective.dev',
    tendermintApi: 'https://devnet-1.tm.injective.dev',
    sentryHttpApi: 'https://devnet-1.lcd.injective.dev',
};
exports.oldEndpointsDevnet2 = {
    indexerApi: 'https://devnet-2.api.injective.dev',
    sentryGrpcApi: 'https://devnet-2.grpc.injective.dev',
    tendermintApi: 'https://devnet-2.tm.injective.dev',
    sentryHttpApi: 'https://devnet-2.lcd.injective.dev',
};
exports.oldEndpointsLocal = {
    indexerApi: 'https://localhost:4444',
    sentryGrpcApi: 'http://localhost:9091',
    tendermintApi: 'http://localhost:9091',
    sentryHttpApi: 'http://localhost:9091',
};
