// @wormhole-foundation/sdk-definitions/dist/cjs/payloads/relayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deviveryInstructionLayout = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const layout_items_1 = require("../layout-items");
const vaa_1 = require("../vaa");
const encodedExecutionInfoItem = {
    binary: "bytes",
    layout: [
        { name: "size", binary: "uint", size: 4, custom: 3 * 32, omit: true },
        { name: "waste", binary: "uint", size: 31, custom: 0n, omit: true },
        { name: "version", binary: "uint", size: 1, custom: 0, omit: true },
        { name: "gasLimit", ...layout_items_1.amountItem },
        { name: "targetChainRefundPerGasUnused", ...layout_items_1.amountItem },
    ],
};
const addressChainItem = {
    binary: "bytes",
    layout: [
        { name: "chain", ...(0, layout_items_1.chainItem)() },
        { name: "address", ...layout_items_1.universalAddressItem },
    ],
};
const vaaKeyLayout = [
    { name: "chain", ...(0, layout_items_1.chainItem)() },
    { name: "emitterAddress", ...layout_items_1.universalAddressItem },
    { name: "sequence", ...layout_items_1.sequenceItem },
];
const cctpKeyLayout = [
    { name: "size", binary: "uint", size: 4, custom: 12, omit: true },
    { name: "domain", ...layout_items_1.circleDomainItem },
    { name: "nonce", ...layout_items_1.circleNonceItem },
];
const messageKeySwitchLayout = {
    binary: "switch",
    idSize: 1,
    idTag: "keyType",
    layouts: [
        [[1, "VAA"], vaaKeyLayout],
        [[2, "CCTP"], cctpKeyLayout],
    ],
};
const deviveryInstructionLayout = (customPayload) => [
    (0, layout_items_1.payloadIdItem)(1),
    { name: "target", ...addressChainItem },
    (0, sdk_base_1.customizableBytes)({ name: "payload", lengthSize: 4 }, customPayload),
    { name: "requestedReceiverValue", ...layout_items_1.amountItem },
    { name: "extraReceiverValue", ...layout_items_1.amountItem },
    { name: "executionInfo", ...encodedExecutionInfoItem },
    { name: "refund", ...addressChainItem },
    { name: "refundDeliveryProvider", ...layout_items_1.universalAddressItem },
    { name: "sourceDeliveryProvider", ...layout_items_1.universalAddressItem },
    { name: "senderAddress", ...layout_items_1.universalAddressItem },
    { name: "messageKeys", binary: "array", lengthSize: 1, layout: messageKeySwitchLayout },
];
exports.deviveryInstructionLayout = deviveryInstructionLayout;
const namedPayloads = [
    ["DeliveryInstruction", (0, exports.deviveryInstructionLayout)()],
    ["RedeliveryInstruction", [
            (0, layout_items_1.payloadIdItem)(2),
            { name: "deliveryVaaKey", binary: "bytes", layout: vaaKeyLayout },
            { name: "targetChain", ...(0, layout_items_1.chainItem)() },
            { name: "newRequestedReceiverValue", ...layout_items_1.amountItem },
            { name: "newEncodedExecutionInfo", ...encodedExecutionInfoItem },
            { name: "newSourceDeliveryProvider", ...layout_items_1.universalAddressItem },
            { name: "newSenderAddress", ...layout_items_1.universalAddressItem },
        ],
    ],
];
(0, vaa_1.registerPayloadTypes)("Relayer", namedPayloads);
