// @certusone/wormhole-sdk/lib/esm/ethers-contracts/factories/WormholeRelayerGovernance__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "uint16",
                name: "chainId",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "registeredWormholeRelayerContract",
                type: "bytes32",
            },
        ],
        name: "ChainAlreadyRegistered",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "failure",
                type: "bytes",
            },
        ],
        name: "ContractUpgradeFailed",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "hash",
                type: "bytes32",
            },
        ],
        name: "GovernanceActionAlreadyConsumed",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "defaultDeliveryProvider",
                type: "bytes32",
            },
        ],
        name: "InvalidDefaultDeliveryProvider",
        type: "error",
    },
    {
        inputs: [],
        name: "InvalidFork",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "parsed",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "expected",
                type: "uint16",
            },
        ],
        name: "InvalidGovernanceChainId",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "parsed",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "expected",
                type: "bytes32",
            },
        ],
        name: "InvalidGovernanceContract",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "reason",
                type: "string",
            },
        ],
        name: "InvalidGovernanceVM",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "parsed",
                type: "uint8",
            },
            {
                internalType: "uint8",
                name: "expected",
                type: "uint8",
            },
        ],
        name: "InvalidPayloadAction",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "parsed",
                type: "uint16",
            },
            {
                internalType: "uint16",
                name: "expected",
                type: "uint16",
            },
        ],
        name: "InvalidPayloadChainId",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "received",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "expected",
                type: "uint256",
            },
        ],
        name: "InvalidPayloadLength",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "parsed",
                type: "bytes32",
            },
            {
                internalType: "bytes32",
                name: "expected",
                type: "bytes32",
            },
        ],
        name: "InvalidPayloadModule",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        name: "NotAnEvmAddress",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "previousAdmin",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address",
                name: "newAdmin",
                type: "address",
            },
        ],
        name: "AdminChanged",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "beacon",
                type: "address",
            },
        ],
        name: "BeaconUpgraded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "oldContract",
                type: "address",
            },
            {
                indexed: true,
                internalType: "address",
                name: "newContract",
                type: "address",
            },
        ],
        name: "ContractUpgraded",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
            {
                indexed: false,
                internalType: "LocalNative",
                name: "deliveryQuote",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
        ],
        name: "SendEvent",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "implementation",
                type: "address",
            },
        ],
        name: "Upgraded",
        type: "event",
    },
    {
        inputs: [],
        name: "checkAndExecuteUpgradeMigration",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliveryAttempted",
        outputs: [
            {
                internalType: "bool",
                name: "attempted",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliveryFailureBlock",
        outputs: [
            {
                internalType: "uint256",
                name: "blockNumber",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliverySuccessBlock",
        outputs: [
            {
                internalType: "uint256",
                name: "blockNumber",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "chainId",
                type: "uint16",
            },
        ],
        name: "getRegisteredWormholeRelayerContract",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "encodedVm",
                type: "bytes",
            },
        ],
        name: "registerWormholeRelayerContract",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "encodedVm",
                type: "bytes",
            },
        ],
        name: "setDefaultDeliveryProvider",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "encodedVm",
                type: "bytes",
            },
        ],
        name: "submitContractUpgrade",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class WormholeRelayerGovernance__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
WormholeRelayerGovernance__factory.abi = _abi;
