// @wormhole-foundation/connect-sdk/dist/cjs/routes/cctp/manual.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CCTPRoute = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const sdk_definitions_1 = require("@wormhole-foundation/sdk-definitions");
const common_1 = require("../../common");
const cctpTransfer_1 = require("../../protocols/cctpTransfer");
const types_1 = require("../../types");
const route_1 = require("../route");
const wormhole_1 = require("../../wormhole");
class CCTPRoute extends route_1.ManualRoute {
    static meta = {
        name: "ManualCCTP",
    };
    static supportedNetworks() {
        return ["Mainnet", "Testnet"];
    }
    // get the list of chains this route supports
    static supportedChains(network) {
        if (sdk_base_1.contracts.circleContractChains.has(network)) {
            return sdk_base_1.contracts.circleContractChains.get(network);
        }
        return [];
    }
    // get the list of source tokens that are possible to send
    static async supportedSourceTokens(fromChain) {
        const { network, chain } = fromChain;
        if (!sdk_base_1.circle.usdcContract.has(network, chain))
            return [];
        return [wormhole_1.Wormhole.chainAddress(chain, sdk_base_1.circle.usdcContract.get(network, chain))];
    }
    // get the liist of destination tokens that may be recieved on the destination chain
    static async supportedDestinationTokens(sourceToken, fromChain, toChain) {
        const { network, chain } = toChain;
        if (!sdk_base_1.circle.usdcContract.has(network, chain))
            return [];
        return [wormhole_1.Wormhole.chainAddress(chain, sdk_base_1.circle.usdcContract.get(network, chain))];
    }
    static isProtocolSupported(chain) {
        return chain.supportsCircleBridge();
    }
    getDefaultOptions() {
        return {
            payload: undefined,
        };
    }
    async validate(params) {
        const amount = this.request.parseAmount(params.amount);
        const validatedParams = {
            normalizedParams: {
                amount,
            },
            options: params.options ?? this.getDefaultOptions(),
            ...params,
        };
        return { valid: true, params: validatedParams };
    }
    async quote(params) {
        try {
            return this.request.displayQuote(await cctpTransfer_1.CircleTransfer.quoteTransfer(this.request.fromChain, this.request.toChain, this.toTransferDetails(params)), params);
        }
        catch (e) {
            return {
                success: false,
                error: e,
            };
        }
    }
    async initiate(signer, quote) {
        const { params } = quote;
        let transfer = this.toTransferDetails(params);
        let txids = await cctpTransfer_1.CircleTransfer.transfer(this.request.fromChain, transfer, signer);
        const msg = await cctpTransfer_1.CircleTransfer.getTransferMessage(this.request.fromChain, txids[txids.length - 1].txid);
        return {
            from: transfer.from.chain,
            to: transfer.to.chain,
            state: types_1.TransferState.SourceFinalized,
            originTxs: txids,
            attestation: { id: msg.id, attestation: { message: msg.message } },
        };
    }
    async complete(signer, receipt) {
        if (!(0, types_1.isAttested)(receipt))
            throw new Error("The source must be finalized in order to complete the transfer");
        const { id, attestation: att } = receipt.attestation;
        if (sdk_definitions_1.CircleBridge.isCircleAttestation(att)) {
            const { message, attestation } = att;
            if (!attestation)
                throw new Error(`No Circle attestation for ${id}`);
            let cb = await this.request.toChain.getCircleBridge();
            let xfer = cb.redeem(this.request.to.address, message, attestation);
            return await (0, common_1.signSendWait)(this.request.toChain, xfer, signer);
        }
        else {
            //
            return [];
        }
    }
    async *track(receipt, timeout) {
        yield* cctpTransfer_1.CircleTransfer.track(this.wh, receipt, timeout, this.request.fromChain, this.request.toChain);
    }
    toTransferDetails(params) {
        return {
            from: this.request.from,
            to: this.request.to,
            amount: sdk_base_1.amount.units(params.normalizedParams.amount),
            automatic: false,
            ...params.options,
        };
    }
}
exports.CCTPRoute = CCTPRoute;
