// @injectivelabs/sdk-ts/dist/esm/client/chain/rest/ChainRestTendermintApi.js
import { HttpRequestException, UnspecifiedErrorCode, } from '@injectivelabs/exceptions';
import BaseRestConsumer from '../../BaseRestConsumer';
import { ChainModule } from '../types';
/**
 * @category Chain Rest API
 */
export class ChainRestTendermintApi extends BaseRestConsumer {
    async fetchLatestBlock() {
        try {
            const response = (await this.get(`cosmos/base/tendermint/v1beta1/blocks/latest`));
            return response.data.block;
        }
        catch (e) {
            if (e instanceof HttpRequestException) {
                throw e;
            }
            throw new HttpRequestException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                contextModule: ChainModule.Tendermint,
            });
        }
    }
    async fetchNodeInfo() {
        try {
            const response = (await this.get(`cosmos/base/tendermint/v1beta1/node_info`));
            return {
                nodeInfo: response.data.default_node_info,
                applicationVersion: response.data.application_version,
            };
        }
        catch (e) {
            if (e instanceof HttpRequestException) {
                throw e;
            }
            throw new HttpRequestException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                contextModule: ChainModule.Tendermint,
            });
        }
    }
}
