// @injectivelabs/sdk-ts/dist/esm/client/chain/rest/ChainRestBankApi.js
import { StatusCodes } from 'http-status-codes';
import { ErrorType, GeneralException, HttpRequestException, UnspecifiedErrorCode, } from '@injectivelabs/exceptions';
import BaseRestConsumer from '../../BaseRestConsumer';
import { ChainModule } from '../types';
/**
 * @category Chain Rest API
 */
export class ChainRestBankApi extends BaseRestConsumer {
    /**
     * Get address's balance
     *
     * @param address address of account to look up
     */
    async fetchBalances(address) {
        try {
            const response = (await this.get(`cosmos/bank/v1beta1/balances/${address}`));
            return response.data;
        }
        catch (e) {
            if (e instanceof HttpRequestException) {
                throw e;
            }
            throw new HttpRequestException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                contextModule: ChainModule.Bank,
            });
        }
    }
    /**
     * Get address's balances
     *
     * @param address address of account to look up
     */
    async fetchBalance(address, denom) {
        try {
            const response = (await this.get(`cosmos/bank/v1beta1/balances/${address}`));
            const balance = response.data.balances.find((balance) => balance.denom === denom);
            if (!balance) {
                throw new GeneralException(new Error(`The ${denom} balance was not found`), {
                    code: StatusCodes.NOT_FOUND,
                    type: ErrorType.NotFoundError,
                });
            }
            return balance;
        }
        catch (e) {
            if (e instanceof HttpRequestException) {
                throw e;
            }
            if (e instanceof GeneralException) {
                throw e;
            }
            throw new HttpRequestException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                contextModule: ChainModule.Bank,
            });
        }
    }
}
