// @certusone/wormhole-sdk/lib/esm/ethers-contracts/factories/IDeliveryProviderTyped.sol/IDeliveryProvider__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "getRewardAddress",
        outputs: [
            {
                internalType: "address payable",
                name: "rewardAddress",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [],
        name: "getSupportedKeys",
        outputs: [
            {
                internalType: "uint256",
                name: "bitmap",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
        ],
        name: "getTargetChainAddress",
        outputs: [
            {
                internalType: "bytes32",
                name: "deliveryProviderAddress",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
        ],
        name: "isChainSupported",
        outputs: [
            {
                internalType: "bool",
                name: "supported",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "keyType",
                type: "uint8",
            },
        ],
        name: "isMessageKeyTypeSupported",
        outputs: [
            {
                internalType: "bool",
                name: "supported",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "LocalNative",
                name: "currentChainAmount",
                type: "uint256",
            },
        ],
        name: "quoteAssetConversion",
        outputs: [
            {
                internalType: "TargetNative",
                name: "targetChainAmount",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionParams",
                type: "bytes",
            },
        ],
        name: "quoteDeliveryPrice",
        outputs: [
            {
                internalType: "LocalNative",
                name: "nativePriceQuote",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionInfo",
                type: "bytes",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
export class IDeliveryProvider__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
IDeliveryProvider__factory.abi = _abi;
