// @across-protocol/across-token/dist/hardhat.config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const dotenv = __importStar(require("dotenv"));
const common_1 = require("@uma/common");
require("@nomiclabs/hardhat-etherscan");
require("@nomiclabs/hardhat-waffle");
require("@typechain/hardhat");
require("hardhat-gas-reporter");
require("solidity-coverage");
require("hardhat-deploy");
dotenv.config();
const solcVersion = "0.8.17";
const mnemonic = (0, common_1.getMnemonic)();
const config = {
    solidity: {
        compilers: [{ version: solcVersion, settings: { optimizer: { enabled: true, runs: 1000000 } } }],
    },
    networks: {
        hardhat: { accounts: { accountsBalance: "1000000000000000000000000" } },
        mainnet: {
            url: (0, common_1.getNodeUrl)("mainnet", true, 1),
            accounts: { mnemonic },
            saveDeployments: true,
            chainId: 1,
        },
        goerli: {
            url: (0, common_1.getNodeUrl)("goerli", true, 5),
            accounts: { mnemonic },
            saveDeployments: true,
            chainId: 5,
        },
    },
    gasReporter: { enabled: process.env.REPORT_GAS !== undefined, currency: "USD" },
    etherscan: {
        apiKey: {
            mainnet: process.env.ETHERSCAN_API_KEY,
            goerli: process.env.ETHERSCAN_API_KEY,
        },
    },
    namedAccounts: { deployer: 0 },
};
exports.default = config;
