// @across-protocol/sdk-v2/dist/esm/caching/Arweave/ArweaveClient.js
import { __awaiter, __generator } from "tslib";
import Arweave from "arweave";
import { ethers } from "ethers";
import { isDefined, jsonReplacerWithBigNumbers, parseWinston, toBN } from "../../utils";
import { is } from "superstruct";
import { ARWEAVE_TAG_APP_NAME } from "../../constants";
var ArweaveClient = /** @class */ (function () {
    function ArweaveClient(arweaveJWT, logger, gatewayURL, protocol, port) {
        if (gatewayURL === void 0) { gatewayURL = "arweave.net"; }
        if (protocol === void 0) { protocol = "https"; }
        if (port === void 0) { port = 443; }
        this.arweaveJWT = arweaveJWT;
        this.logger = logger;
        this.client = new Arweave({
            host: gatewayURL,
            port: port,
            protocol: protocol,
            timeout: 20000,
            logging: false,
        });
        this.logger.info("Arweave client initialized");
    }
    /**
     * Stores an arbitrary record in the Arweave network. The record is stored as a JSON string and uses
     * JSON.stringify to convert the record to a string. The record has all of its big numbers converted
     * to strings for convenience.
     * @param value The value to store
     * @param topicTag An optional topic tag to add to the transaction
     * @returns The transaction ID of the stored value
     * @
     */
    ArweaveClient.prototype.set = function (value, topicTag) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var transaction, result, message, _d, _e;
            var _f;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0: return [4 /*yield*/, this.client.createTransaction({ data: JSON.stringify(value, jsonReplacerWithBigNumbers) }, this.arweaveJWT)];
                    case 1:
                        transaction = _g.sent();
                        // Add tags to the transaction
                        transaction.addTag("Content-Type", "application/json");
                        transaction.addTag("App-Name", ARWEAVE_TAG_APP_NAME);
                        if (isDefined(topicTag)) {
                            transaction.addTag("Topic", topicTag);
                        }
                        // Sign the transaction
                        return [4 /*yield*/, this.client.transactions.sign(transaction, this.arweaveJWT)];
                    case 2:
                        // Sign the transaction
                        _g.sent();
                        return [4 /*yield*/, this.client.transactions.post(transaction)];
                    case 3:
                        result = _g.sent();
                        if (!(result.status !== 200)) return [3 /*break*/, 6];
                        message = (_c = (_b = (_a = result === null || result === void 0 ? void 0 : result.data) === null || _a === void 0 ? void 0 : _a.error) === null || _b === void 0 ? void 0 : _b.msg) !== null && _c !== void 0 ? _c : "Unknown error";
                        _e = (_d = this.logger).error;
                        _f = {
                            at: "ArweaveClient:set",
                            message: message,
                            result: result,
                            txn: transaction.id
                        };
                        return [4 /*yield*/, this.getAddress()];
                    case 4:
                        _f.address = _g.sent();
                        return [4 /*yield*/, this.getBalance()];
                    case 5:
                        _e.apply(_d, [(_f.balance = (_g.sent()).toString(),
                                _f)]);
                        throw new Error(message);
                    case 6:
                        this.logger.debug({
                            at: "ArweaveClient:set",
                            message: "Arweave transaction posted with ".concat(transaction.id),
                        });
                        _g.label = 7;
                    case 7: return [2 /*return*/, transaction.id];
                }
            });
        });
    };
    /**
     * Retrieves a record from the Arweave network. The record is expected to be a JSON string and is
     * parsed using JSON.parse. All numeric strings are converted to big numbers for convenience.
     * @param transactionID The transaction ID of the record to retrieve
     * @param structValidator An optional struct validator to validate the retrieved value. If the value does not match the struct, null is returned.
     * @returns The record if it exists, otherwise null
     */
    ArweaveClient.prototype.get = function (transactionID, validator) {
        return __awaiter(this, void 0, void 0, function () {
            var rawData, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.transactions.getData(transactionID, { decode: true, string: true })];
                    case 1:
                        rawData = _a.sent();
                        if (!rawData) {
                            return [2 /*return*/, null];
                        }
                        data = JSON.parse(typeof rawData === "string" ? rawData : Buffer.from(rawData).toString("utf-8"));
                        // Ensure that the result is successful. If it is not, the retrieved value is not our expected type
                        // but rather a {status: string, statusText: string} object. We can detect that and return null.
                        if (data.status === 400) {
                            return [2 /*return*/, null];
                        }
                        // If the validator does not match the retrieved value, return null and log a warning
                        if (!is(data, validator)) {
                            this.logger.warn("Retrieved value from Arweave does not match the expected type");
                            return [2 /*return*/, null];
                        }
                        return [2 /*return*/, data];
                }
            });
        });
    };
    /**
     * Retrieves a list of records from the Arweave network that have a specific tag.
     * The records are expected to be a JSON string and are pre-filtered by the Across
     * protocol tag, the content-type tag, and this client's address. Furthermore, the
     * records are expected to be an array of the given type and will be discarded if
     * they do not match the given validator.
     * @param tag The tag to filter all the transactions by
     * @param validator The validator to validate the retrieved values
     * @returns The records if they exist, otherwise an empty array
     */
    ArweaveClient.prototype.getByTopic = function (tag, validator) {
        return __awaiter(this, void 0, void 0, function () {
            var transactions, _a, _b, _c, _d, results;
            var _e;
            var _this = this;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _b = (_a = this.client.api).post;
                        _c = ["/graphql"];
                        _e = {};
                        _d = "\n        { \n          transactions (\n            owners: [\"".concat;
                        return [4 /*yield*/, this.getAddress()];
                    case 1: return [4 /*yield*/, _b.apply(_a, _c.concat([(_e.query = _d.apply("\n        { \n          transactions (\n            owners: [\"", [_f.sent(), "\"]\n            tags: [\n              { name: \"App-Name\", values: [\""]).concat(ARWEAVE_TAG_APP_NAME, "\"] },\n              { name: \"Content-Type\", values: [\"application/json\"] },\n              ").concat(tag ? "{ name: \"Topic\", values: [\"".concat(tag, "\"] } ") : "", "\n            ]\n          ) { edges { node { id } } } \n        }"),
                                _e)]))];
                    case 2:
                        transactions = _f.sent();
                        return [4 /*yield*/, Promise.all(transactions.data.data.transactions.edges.map(function (edge) { return __awaiter(_this, void 0, void 0, function () {
                                var data;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.get(edge.node.id, validator)];
                                        case 1:
                                            data = _a.sent();
                                            return [2 /*return*/, isDefined(data)
                                                    ? {
                                                        data: data,
                                                        hash: edge.node.id,
                                                    }
                                                    : null];
                                    }
                                });
                            }); }))];
                    case 3:
                        results = _f.sent();
                        return [2 /*return*/, results.filter(isDefined)];
                }
            });
        });
    };
    /**
     * Retrieves the metadata of a transaction
     * @param transactionID The transaction ID of the record to retrieve
     * @returns The metadata of the transaction if it exists, otherwise null
     */
    ArweaveClient.prototype.getMetadata = function (transactionID) {
        return __awaiter(this, void 0, void 0, function () {
            var transaction, tags;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.client.transactions.get(transactionID)];
                    case 1:
                        transaction = _a.sent();
                        if (!isDefined(transaction)) {
                            return [2 /*return*/, null];
                        }
                        tags = Object.fromEntries(transaction.tags.map(function (tag) { return [
                            tag.get("name", { decode: true, string: true }),
                            tag.get("value", { decode: true, string: true }),
                        ]; }));
                        return [2 /*return*/, {
                                contentType: tags["Content-Type"],
                                appName: tags["App-Name"],
                                topic: tags.Topic,
                            }];
                }
            });
        });
    };
    /**
     * Returns the address of the signer of the JWT
     * @returns The address of the signer in this client
     */
    ArweaveClient.prototype.getAddress = function () {
        return this.client.wallets.jwkToAddress(this.arweaveJWT);
    };
    /**
     * The balance of the signer
     * @returns The balance of the signer in winston units
     */
    ArweaveClient.prototype.getBalance = function () {
        return __awaiter(this, void 0, void 0, function () {
            var address, balanceInFloat, _a, balance, exponent, resultingBN;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getAddress()];
                    case 1:
                        address = _b.sent();
                        return [4 /*yield*/, this.client.wallets.getBalance(address)];
                    case 2:
                        balanceInFloat = _b.sent();
                        // Sometimes the balance is returned in scientific notation, so we need to
                        // convert it to a BigNumber
                        if (balanceInFloat.includes("e")) {
                            _a = balanceInFloat.split("e"), balance = _a[0], exponent = _a[1];
                            resultingBN = ethers.BigNumber.from(balance).mul(toBN(10).pow(exponent.replace("+", "")));
                            return [2 /*return*/, parseWinston(resultingBN.toString())];
                        }
                        else {
                            return [2 /*return*/, parseWinston(balanceInFloat)];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return ArweaveClient;
}());
export { ArweaveClient };
