// @across-protocol/sdk-v2/dist/cjs/utils/BigNumberUtils.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.toBN = exports.toGWei = exports.toWei = exports.bnUint32Max = exports.bnUint256Max = exports.bnOne = exports.bnZero = void 0;
var ethers_1 = require("ethers");
exports.bnZero = (_a = ethers_1.ethers.constants, _a.Zero), exports.bnOne = _a.One, exports.bnUint256Max = _a.MaxUint256;
exports.bnUint32Max = ethers_1.BigNumber.from("0xffffffff");
function toWei(num) {
    return ethers_1.ethers.utils.parseEther(num.toString());
}
exports.toWei = toWei;
function toGWei(num) {
    return ethers_1.ethers.utils.parseUnits(num.toString(), 9);
}
exports.toGWei = toGWei;
var toBN = function (num, rounding) {
    if (rounding === void 0) { rounding = "floor"; }
    if (num.toString().includes(".")) {
        var _a = num.toString().split("."), integer = _a[0], decimal = _a[1];
        var roundUp = decimal.length > 0 && (rounding === "ceil" || (rounding === "round" && parseInt(decimal[0]) >= 5));
        return ethers_1.BigNumber.from(integer).add(roundUp ? 1 : 0);
    }
    return ethers_1.BigNumber.from(num.toString());
};
exports.toBN = toBN;
