// @across-protocol/sdk-v2/dist/cjs/utils/NetworkUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainIsArbitrum = exports.chainIsOPStack = exports.chainIsTestnet = exports.chainIsProd = exports.getNativeTokenSymbol = exports.getNetworkName = void 0;
var PublicNetworks_1 = require("@uma/common/dist/PublicNetworks");
var constants_1 = require("../constants");
var networkIdMap = {
    666: "Hardhat1",
    1337: "Hardhat2",
    421613: "ArbitrumGoerli",
    421614: "ArbitrumSepolia",
    324: "ZkSync",
    280: "ZkSync-Goerli",
    300: "ZKSync-Sepolia",
    8453: "Base",
    84531: "BaseGoerli",
    84532: "BaseSepolia",
    11155111: "EthSepolia",
    11155420: "OptimismSepolia",
};
function getNetworkName(networkId) {
    try {
        var networkName = PublicNetworks_1.PublicNetworks[Number(networkId)].name;
        return networkName.charAt(0).toUpperCase() + networkName.slice(1);
    }
    catch (error) {
        return networkIdMap[Number(networkId)] || "unknown";
    }
}
exports.getNetworkName = getNetworkName;
function getNativeTokenSymbol(chainId) {
    if (chainId.toString() === "137" || chainId.toString() === "80001") {
        return "MATIC";
    }
    return "ETH";
}
exports.getNativeTokenSymbol = getNativeTokenSymbol;
function chainIsProd(chainId) {
    return constants_1.PRODUCTION_CHAIN_IDS.includes(chainId);
}
exports.chainIsProd = chainIsProd;
function chainIsTestnet(chainId) {
    return constants_1.TESTNET_CHAIN_IDS.includes(chainId);
}
exports.chainIsTestnet = chainIsTestnet;
function chainIsOPStack(chainId) {
    return [
        constants_1.CHAIN_IDs.OPTIMISM,
        constants_1.CHAIN_IDs.BASE,
        constants_1.CHAIN_IDs.OPTIMISM_GOERLI,
        constants_1.CHAIN_IDs.BASE_GOERLI,
        constants_1.CHAIN_IDs.OPTIMISM_SEPOLIA,
        constants_1.CHAIN_IDs.BASE_SEPOLIA,
    ].includes(chainId);
}
exports.chainIsOPStack = chainIsOPStack;
function chainIsArbitrum(chainId) {
    return [constants_1.CHAIN_IDs.ARBITRUM, constants_1.CHAIN_IDs.ARBITRUM_GOERLI, constants_1.CHAIN_IDs.ARBITRUM_SEPOLIA].includes(chainId);
}
exports.chainIsArbitrum = chainIsArbitrum;
