// @across-protocol/sdk-v2/dist/cjs/utils/V3Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSlowFillLeafLpFeePct = exports.getSlowFillLeafChainId = exports.getRelayDataOutputAmount = exports.getRelayDataOutputToken = exports.getTotalFilledAmount = exports.getFillAmount = exports.getFillOutputAmount = exports.getFillOutputToken = exports.getDepositOutputAmount = exports.getDepositInputAmount = exports.getDepositOutputToken = exports.getDepositInputToken = exports.isV3SlowFillLeaf = exports.isV2SlowFillLeaf = exports.isSlowFill = exports.isV3RelayData = exports.isV2RelayData = exports.isV3Fill = exports.isV2Fill = exports.isV3SpeedUp = exports.isV2SpeedUp = exports.isV3Deposit = exports.isV2Deposit = exports.isType = exports.isV3 = exports.V3_MIN_CONFIG_STORE_VERSION = void 0;
var interfaces_1 = require("../interfaces");
var common_1 = require("./common");
exports.V3_MIN_CONFIG_STORE_VERSION = 3;
function isV3(version) {
    return version >= exports.V3_MIN_CONFIG_STORE_VERSION;
}
exports.isV3 = isV3;
function isType(input, key) {
    return input[key] !== undefined;
}
exports.isType = isType;
function unsafeIsType(input, key) {
    return input[key] !== undefined;
}
function isV2Deposit(deposit) {
    return unsafeIsType(deposit, "originToken");
}
exports.isV2Deposit = isV2Deposit;
function isV3Deposit(deposit) {
    return unsafeIsType(deposit, "inputToken");
}
exports.isV3Deposit = isV3Deposit;
function isV2SpeedUp(speedUp) {
    return unsafeIsType(speedUp, "newRelayerFeePct");
}
exports.isV2SpeedUp = isV2SpeedUp;
function isV3SpeedUp(speedUp) {
    return unsafeIsType(speedUp, "updatedOutputAmount");
}
exports.isV3SpeedUp = isV3SpeedUp;
function isV2Fill(fill) {
    return unsafeIsType(fill, "destinationToken");
}
exports.isV2Fill = isV2Fill;
function isV3Fill(fill) {
    return unsafeIsType(fill, "inputToken");
}
exports.isV3Fill = isV3Fill;
function isV2RelayData(relayData) {
    return unsafeIsType(relayData, "destinationToken");
}
exports.isV2RelayData = isV2RelayData;
function isV3RelayData(relayData) {
    return unsafeIsType(relayData, "outputToken");
}
exports.isV3RelayData = isV3RelayData;
function isSlowFill(fill) {
    return isV2Fill(fill) ? fill.updatableRelayData.isSlowRelay : fill.relayExecutionInfo.fillType === interfaces_1.FillType.SlowFill;
}
exports.isSlowFill = isSlowFill;
function isV2SlowFillLeaf(slowFillLeaf) {
    return unsafeIsType(slowFillLeaf, "payoutAdjustmentPct");
}
exports.isV2SlowFillLeaf = isV2SlowFillLeaf;
function isV3SlowFillLeaf(slowFillLeaf) {
    return unsafeIsType(slowFillLeaf, "updatedOutputAmount");
}
exports.isV3SlowFillLeaf = isV3SlowFillLeaf;
function getDepositInputToken(deposit) {
    return isV2Deposit(deposit) ? deposit.originToken : deposit.inputToken;
}
exports.getDepositInputToken = getDepositInputToken;
function getDepositOutputToken(deposit) {
    return unsafeIsType(deposit, "destinationToken") ? deposit.destinationToken : deposit.outputToken;
}
exports.getDepositOutputToken = getDepositOutputToken;
function getDepositInputAmount(deposit) {
    return unsafeIsType(deposit, "amount") ? deposit.amount : deposit.inputAmount;
}
exports.getDepositInputAmount = getDepositInputAmount;
function getDepositOutputAmount(deposit) {
    return unsafeIsType(deposit, "amount") ? deposit.amount : deposit.outputAmount;
}
exports.getDepositOutputAmount = getDepositOutputAmount;
function getFillOutputToken(fill) {
    return unsafeIsType(fill, "destinationToken") ? fill.destinationToken : fill.outputToken;
}
exports.getFillOutputToken = getFillOutputToken;
function getFillOutputAmount(fill) {
    return unsafeIsType(fill, "amount") ? fill.amount : fill.outputAmount;
}
exports.getFillOutputAmount = getFillOutputAmount;
function getFillAmount(fill) {
    return unsafeIsType(fill, "fillAmount") ? fill.fillAmount : fill.outputAmount;
}
exports.getFillAmount = getFillAmount;
function getTotalFilledAmount(fill) {
    return unsafeIsType(fill, "totalFilledAmount") ? fill.totalFilledAmount : fill.outputAmount;
}
exports.getTotalFilledAmount = getTotalFilledAmount;
function getRelayDataOutputToken(relayData) {
    return isV2RelayData(relayData) ? relayData.destinationToken : relayData.outputToken;
}
exports.getRelayDataOutputToken = getRelayDataOutputToken;
function getRelayDataOutputAmount(relayData) {
    return unsafeIsType(relayData, "amount") ? relayData.amount : relayData.outputAmount;
}
exports.getRelayDataOutputAmount = getRelayDataOutputAmount;
function getSlowFillLeafChainId(leaf) {
    return unsafeIsType(leaf, "chainId") ? leaf.chainId : leaf.relayData.destinationChainId;
}
exports.getSlowFillLeafChainId = getSlowFillLeafChainId;
function getSlowFillLeafLpFeePct(leaf) {
    return unsafeIsType(leaf, "updatedOutputAmount")
        ? leaf.relayData.inputAmount.sub(leaf.updatedOutputAmount).mul(common_1.fixedPointAdjustment).div(leaf.relayData.inputAmount)
        : leaf.relayData.realizedLpFeePct;
}
exports.getSlowFillLeafLpFeePct = getSlowFillLeafLpFeePct;
