// @certusone/wormhole-sdk-proto-web/lib/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.spy = exports.publicrpc = exports.node = exports.gossip = exports.descriptor = exports.http = exports.annotations = void 0;
exports.annotations = __importStar(require("./google/api/annotations"));
exports.http = __importStar(require("./google/api/http"));
exports.descriptor = __importStar(require("./google/protobuf/descriptor"));
exports.gossip = __importStar(require("./gossip/v1/gossip"));
exports.node = __importStar(require("./node/v1/node"));
exports.publicrpc = __importStar(require("./publicrpc/v1/publicrpc"));
exports.spy = __importStar(require("./spy/v1/spy"));
