// @wormhole-foundation/connect-sdk/dist/cjs/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_NETWORK = exports.applyOverrides = exports.networkPlatformConfigs = exports.CONFIG = exports.DEFAULT_TASK_TIMEOUT = void 0;
const sdk_base_1 = require("@wormhole-foundation/sdk-base");
const sdk_definitions_1 = require("@wormhole-foundation/sdk-definitions");
exports.DEFAULT_TASK_TIMEOUT = 60 * 1000; // 1 minute in milliseconds
exports.CONFIG = {
    Mainnet: {
        api: "https://api.wormholescan.io",
        circleAPI: sdk_base_1.circle.circleAPI("Mainnet"),
        chains: (0, sdk_definitions_1.buildConfig)("Mainnet"),
    },
    Testnet: {
        api: "https://api.testnet.wormholescan.io",
        circleAPI: sdk_base_1.circle.circleAPI("Testnet"),
        chains: (0, sdk_definitions_1.buildConfig)("Testnet"),
    },
    Devnet: {
        api: "http://guardian:7071",
        circleAPI: "",
        chains: (0, sdk_definitions_1.buildConfig)("Devnet"),
    },
};
function networkPlatformConfigs(network, platform) {
    return Object.fromEntries(Object.entries(exports.CONFIG[network].chains).filter(([_, c]) => {
        return c.platform == platform;
    }));
}
exports.networkPlatformConfigs = networkPlatformConfigs;
// Apply any overrides to the base config
function applyOverrides(network, overrides) {
    let base = exports.CONFIG[network];
    if (!overrides)
        return base;
    // recurse through the overrides and apply them to the base config
    function override(base, overrides) {
        for (const [key, value] of Object.entries(overrides)) {
            if (typeof value === "object" && !Array.isArray(value)) {
                base[key] = override(base[key], value);
            }
            else {
                base[key] = value;
            }
        }
        return base;
    }
    return override(base, overrides);
}
exports.applyOverrides = applyOverrides;
const inNode = typeof process !== "undefined";
exports.DEFAULT_NETWORK = (inNode && process.env["NETWORK"]) || "Testnet";
