// @wormhole-foundation/connect-sdk/dist/cjs/wormholeTransfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSourceFinalized = exports.isSourceInitiated = exports.isAttested = exports.TransferState = void 0;
// Transfer state machine states
var TransferState;
(function (TransferState) {
    TransferState[TransferState["Failed"] = -1] = "Failed";
    TransferState[TransferState["Created"] = 0] = "Created";
    TransferState[TransferState["SourceInitiated"] = 1] = "SourceInitiated";
    TransferState[TransferState["SourceFinalized"] = 2] = "SourceFinalized";
    TransferState[TransferState["Attested"] = 3] = "Attested";
    TransferState[TransferState["DestinationInitiated"] = 4] = "DestinationInitiated";
    TransferState[TransferState["DestinationFinalized"] = 5] = "DestinationFinalized";
})(TransferState || (exports.TransferState = TransferState = {}));
function isAttested(receipt) {
    return receipt.state === TransferState.Attested;
}
exports.isAttested = isAttested;
function isSourceInitiated(receipt) {
    return receipt.state === TransferState.SourceInitiated;
}
exports.isSourceInitiated = isSourceInitiated;
function isSourceFinalized(receipt) {
    return receipt.state === TransferState.SourceFinalized;
}
exports.isSourceFinalized = isSourceFinalized;
