// @certusone/wormhole-sdk/lib/cjs/solana/nftBridge/accounts/wrapped.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrappedMeta = exports.getWrappedMeta = exports.deriveWrappedMintKey = exports.deriveWrappedMetaKey = void 0;
const anchor_1 = require("@project-serum/anchor");
const utils_1 = require("../../../utils");
const utils_2 = require("../../utils");
const tokenBridge_1 = require("../../tokenBridge");
var tokenBridge_2 = require("../../tokenBridge");
Object.defineProperty(exports, "deriveWrappedMetaKey", { enumerable: true, get: function () { return tokenBridge_2.deriveWrappedMetaKey; } });
function deriveWrappedMintKey(tokenBridgeProgramId, tokenChain, tokenAddress, tokenId) {
    if (tokenChain == utils_1.CHAIN_ID_SOLANA) {
        throw new Error("tokenChain == CHAIN_ID_SOLANA does not have wrapped mint key");
    }
    if (typeof tokenAddress == "string") {
        tokenAddress = utils_1.tryNativeToUint8Array(tokenAddress, tokenChain);
    }
    return utils_2.deriveAddress([
        Buffer.from("wrapped"),
        (() => {
            const buf = Buffer.alloc(2);
            buf.writeUInt16BE(tokenChain);
            return buf;
        })(),
        tokenAddress,
        new anchor_1.BN(tokenId.toString()).toArrayLike(Buffer, "be", 32),
    ], tokenBridgeProgramId);
}
exports.deriveWrappedMintKey = deriveWrappedMintKey;
function getWrappedMeta(connection, tokenBridgeProgramId, mint, commitment) {
    return __awaiter(this, void 0, void 0, function* () {
        return connection
            .getAccountInfo(tokenBridge_1.deriveWrappedMetaKey(tokenBridgeProgramId, mint), commitment)
            .then((info) => WrappedMeta.deserialize(utils_2.getAccountData(info)));
    });
}
exports.getWrappedMeta = getWrappedMeta;
class WrappedMeta {
    constructor(chain, tokenAddress, tokenId) {
        this.chain = chain;
        this.tokenAddress = tokenAddress;
        this.tokenId = tokenId;
    }
    static deserialize(data) {
        if (data.length != 66) {
            throw new Error("data.length != 66");
        }
        const chain = data.readUInt16LE(0);
        const tokenAddress = data.subarray(2, 34);
        const tokenId = BigInt(new anchor_1.BN(data.subarray(34, 66), undefined, "le").toString());
        return new WrappedMeta(chain, tokenAddress, tokenId);
    }
}
exports.WrappedMeta = WrappedMeta;
