// @certusone/wormhole-sdk/lib/cjs/ethers-contracts/factories/BytesParsing__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytesParsing__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "uint256",
                name: "offset",
                type: "uint256",
            },
            {
                internalType: "uint256",
                name: "length",
                type: "uint256",
            },
        ],
        name: "OutOfBounds",
        type: "error",
    },
];
const _bytecode = "0x60808060405234601757603a9081601d823930815050f35b600080fdfe600080fdfea26469706673582212206aa98ed59fb413b7935f4075e15e67a4501ca5fcdfe822d7613036a9a2363bc164736f6c63430008130033";
const isSuperArgs = (xs) => xs.length > 1;
class BytesParsing__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    attach(address) {
        return super.attach(address);
    }
    connect(signer) {
        return super.connect(signer);
    }
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.BytesParsing__factory = BytesParsing__factory;
BytesParsing__factory.bytecode = _bytecode;
BytesParsing__factory.abi = _abi;
