// @certusone/wormhole-sdk/lib/cjs/ethers-contracts/factories/IWormholeReceiver__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IWormholeReceiver__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "bytes[]",
                name: "additionalMessages",
                type: "bytes[]",
            },
            {
                internalType: "bytes32",
                name: "sourceAddress",
                type: "bytes32",
            },
            {
                internalType: "uint16",
                name: "sourceChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "receiveWormholeMessages",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
];
class IWormholeReceiver__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IWormholeReceiver__factory = IWormholeReceiver__factory;
IWormholeReceiver__factory.abi = _abi;
