// @certusone/wormhole-sdk/lib/cjs/ethers-contracts/factories/WormholeRelayerSend__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WormholeRelayerSend__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "DeliveryProviderCannotReceivePayment",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "keyType",
                type: "uint8",
            },
        ],
        name: "DeliveryProviderDoesNotSupportMessageKeyType",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "address",
                name: "relayer",
                type: "address",
            },
            {
                internalType: "uint16",
                name: "chainId",
                type: "uint16",
            },
        ],
        name: "DeliveryProviderDoesNotSupportTargetChain",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "LocalNative",
                name: "msgValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "totalFee",
                type: "uint256",
            },
        ],
        name: "InvalidMsgValue",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint256",
                name: "numMessageKeys",
                type: "uint256",
            },
        ],
        name: "TooManyMessageKeys",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "uint8",
                name: "version",
                type: "uint8",
            },
            {
                internalType: "uint8",
                name: "expectedVersion",
                type: "uint8",
            },
        ],
        name: "UnexpectedExecutionInfoVersion",
        type: "error",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
            {
                indexed: false,
                internalType: "LocalNative",
                name: "deliveryQuote",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
        ],
        name: "SendEvent",
        type: "event",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliveryAttempted",
        outputs: [
            {
                internalType: "bool",
                name: "attempted",
                type: "bool",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliveryFailureBlock",
        outputs: [
            {
                internalType: "uint256",
                name: "blockNumber",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes32",
                name: "deliveryHash",
                type: "bytes32",
            },
        ],
        name: "deliverySuccessBlock",
        outputs: [
            {
                internalType: "uint256",
                name: "blockNumber",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "targetAddress",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionParameters",
                type: "bytes",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "refundAddress",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "forward",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
        ],
        name: "forwardPayloadToEvm",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "refundAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "forwardToEvm",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
        ],
        name: "forwardVaasToEvm",
        outputs: [],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [],
        name: "getDefaultDeliveryProvider",
        outputs: [
            {
                internalType: "address",
                name: "deliveryProvider",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "chainId",
                type: "uint16",
            },
        ],
        name: "getRegisteredWormholeRelayerContract",
        outputs: [
            {
                internalType: "bytes32",
                name: "",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionParameters",
                type: "bytes",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
        ],
        name: "quoteDeliveryPrice",
        outputs: [
            {
                internalType: "LocalNative",
                name: "nativePriceQuote",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionInfo",
                type: "bytes",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
        ],
        name: "quoteEVMDeliveryPrice",
        outputs: [
            {
                internalType: "LocalNative",
                name: "nativePriceQuote",
                type: "uint256",
            },
            {
                internalType: "GasPrice",
                name: "targetChainRefundPerGasUnused",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
        ],
        name: "quoteEVMDeliveryPrice",
        outputs: [
            {
                internalType: "LocalNative",
                name: "nativePriceQuote",
                type: "uint256",
            },
            {
                internalType: "GasPrice",
                name: "targetChainRefundPerGasUnused",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "LocalNative",
                name: "currentChainAmount",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
        ],
        name: "quoteNativeForChain",
        outputs: [
            {
                internalType: "TargetNative",
                name: "targetChainAmount",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey",
                name: "deliveryVaaKey",
                type: "tuple",
            },
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "newReceiverValue",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "newEncodedExecutionParameters",
                type: "bytes",
            },
            {
                internalType: "address",
                name: "newDeliveryProviderAddress",
                type: "address",
            },
        ],
        name: "resend",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey",
                name: "deliveryVaaKey",
                type: "tuple",
            },
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "TargetNative",
                name: "newReceiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "newGasLimit",
                type: "uint256",
            },
            {
                internalType: "address",
                name: "newDeliveryProviderAddress",
                type: "address",
            },
        ],
        name: "resendToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "targetAddress",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionParameters",
                type: "bytes",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "refundAddress",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "send",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "targetAddress",
                type: "bytes32",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
            {
                internalType: "bytes",
                name: "encodedExecutionParameters",
                type: "bytes",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "bytes32",
                name: "refundAddress",
                type: "bytes32",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "keyType",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes",
                        name: "encodedKey",
                        type: "bytes",
                    },
                ],
                internalType: "struct MessageKey[]",
                name: "messageKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "send",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "refundAddress",
                type: "address",
            },
        ],
        name: "sendPayloadToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
        ],
        name: "sendPayloadToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "refundAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "sendToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "LocalNative",
                name: "paymentForExtraReceiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "refundAddress",
                type: "address",
            },
            {
                internalType: "address",
                name: "deliveryProviderAddress",
                type: "address",
            },
            {
                components: [
                    {
                        internalType: "uint8",
                        name: "keyType",
                        type: "uint8",
                    },
                    {
                        internalType: "bytes",
                        name: "encodedKey",
                        type: "bytes",
                    },
                ],
                internalType: "struct MessageKey[]",
                name: "messageKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint8",
                name: "consistencyLevel",
                type: "uint8",
            },
        ],
        name: "sendToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
        ],
        name: "sendVaasToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint16",
                name: "targetChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "targetAddress",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "payload",
                type: "bytes",
            },
            {
                internalType: "TargetNative",
                name: "receiverValue",
                type: "uint256",
            },
            {
                internalType: "Gas",
                name: "gasLimit",
                type: "uint256",
            },
            {
                components: [
                    {
                        internalType: "uint16",
                        name: "chainId",
                        type: "uint16",
                    },
                    {
                        internalType: "bytes32",
                        name: "emitterAddress",
                        type: "bytes32",
                    },
                    {
                        internalType: "uint64",
                        name: "sequence",
                        type: "uint64",
                    },
                ],
                internalType: "struct VaaKey[]",
                name: "vaaKeys",
                type: "tuple[]",
            },
            {
                internalType: "uint16",
                name: "refundChain",
                type: "uint16",
            },
            {
                internalType: "address",
                name: "refundAddress",
                type: "address",
            },
        ],
        name: "sendVaasToEvm",
        outputs: [
            {
                internalType: "uint64",
                name: "sequence",
                type: "uint64",
            },
        ],
        stateMutability: "payable",
        type: "function",
    },
];
class WormholeRelayerSend__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.WormholeRelayerSend__factory = WormholeRelayerSend__factory;
WormholeRelayerSend__factory.abi = _abi;
