// @across-protocol/sdk-v2/dist/esm/apiClient/abstractClient.js
/**
 * AbstractApiClient is an abstract class that defines the interface for the API client.
 */
var AbstractApiClient = /** @class */ (function () {
    /**
     * Creates a new instance of AbstractApiClient.
     * @param serverlessApiUrl The URL of the serverless API.
     * @param scraperApiUrl The URL of the scraper API.
     */
    function AbstractApiClient(serverlessApiUrl, scraperApiUrl) {
        this.serverlessApiUrl = serverlessApiUrl;
        this.scraperApiUrl = scraperApiUrl;
    }
    /**
     * Returns the URL of the serverless API.
     * @returns The URL of the serverless API.
     */
    AbstractApiClient.prototype.getServerlessApiUrl = function () {
        return this.serverlessApiUrl;
    };
    /**
     * Returns the URL of the scraper API.
     * @returns The URL of the scraper API.
     */
    AbstractApiClient.prototype.getScraperApiUrl = function () {
        return this.scraperApiUrl;
    };
    return AbstractApiClient;
}());
export default AbstractApiClient;
