// @uma/contracts-frontend/dist/typechain/@across-protocol/contracts/ethers/factories/BridgePoolInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "newAdmin",
                type: "address",
            },
        ],
        name: "changeAdmin",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [],
        name: "l1Token",
        outputs: [
            {
                internalType: "contract IERC20",
                name: "",
                type: "address",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "uint64",
                name: "_newLpFeeRatePerSecond",
                type: "uint64",
            },
        ],
        name: "setLpFeeRatePerSecond",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bool",
                name: "_relaysEnabled",
                type: "bool",
            },
        ],
        name: "setRelaysEnabled",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class BridgePoolInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
BridgePoolInterface__factory.abi = _abi;
