// @uma/contracts-frontend/dist/typechain/@across-protocol/contracts/ethers/factories/ParentMessengerInterface__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        inputs: [],
        name: "getL1CallValue",
        outputs: [
            {
                internalType: "uint256",
                name: "",
                type: "uint256",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
    {
        inputs: [
            {
                internalType: "bytes",
                name: "data",
                type: "bytes",
            },
        ],
        name: "sendMessageToChild",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function",
    },
];
export class ParentMessengerInterface__factory {
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
ParentMessengerInterface__factory.abi = _abi;
