// @across-protocol/sdk-v2/dist/cjs/clients/BaseAbstractClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseAbstractClient = void 0;
var utils_1 = require("../utils");
var BaseAbstractClient = (function () {
    function BaseAbstractClient(cachingMechanism) {
        this.cachingMechanism = cachingMechanism;
        this.firstBlockToSearch = 0;
        this.latestBlockSearched = 0;
        this._isUpdated = false;
    }
    Object.defineProperty(BaseAbstractClient.prototype, "isUpdated", {
        get: function () {
            return this._isUpdated;
        },
        set: function (value) {
            if (this._isUpdated === true && !value) {
                throw new Error("Cannot set isUpdated to false once it is true");
            }
            this._isUpdated = value;
        },
        enumerable: false,
        configurable: true
    });
    BaseAbstractClient.prototype.assertUpdated = function () {
        if (!this.isUpdated) {
            throw new Error("Client not updated");
        }
    };
    BaseAbstractClient.prototype.hasCachingMechanism = function () {
        return (0, utils_1.isDefined)(this.cachingMechanism);
    };
    return BaseAbstractClient;
}());
exports.BaseAbstractClient = BaseAbstractClient;
