// @across-protocol/sdk-v2/dist/cjs/clients/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mocks = exports.SpokePoolClient = exports.HubPoolClient = exports.AcrossConfigStoreClient = exports.GLOBAL_CONFIG_STORE_KEYS = exports.DEFAULT_CONFIG_STORE_VERSION = void 0;
var tslib_1 = require("tslib");
var AcrossConfigStoreClient_1 = require("./AcrossConfigStoreClient");
Object.defineProperty(exports, "DEFAULT_CONFIG_STORE_VERSION", { enumerable: true, get: function () { return AcrossConfigStoreClient_1.DEFAULT_CONFIG_STORE_VERSION; } });
Object.defineProperty(exports, "GLOBAL_CONFIG_STORE_KEYS", { enumerable: true, get: function () { return AcrossConfigStoreClient_1.GLOBAL_CONFIG_STORE_KEYS; } });
Object.defineProperty(exports, "AcrossConfigStoreClient", { enumerable: true, get: function () { return AcrossConfigStoreClient_1.AcrossConfigStoreClient; } });
var HubPoolClient_1 = require("./HubPoolClient");
Object.defineProperty(exports, "HubPoolClient", { enumerable: true, get: function () { return HubPoolClient_1.HubPoolClient; } });
var SpokePoolClient_1 = require("./SpokePoolClient");
Object.defineProperty(exports, "SpokePoolClient", { enumerable: true, get: function () { return SpokePoolClient_1.SpokePoolClient; } });
exports.mocks = tslib_1.__importStar(require("./mocks"));
