// @certusone/wormhole-sdk/lib/cjs/mock/nftBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockSolanaNftBridge = exports.MockEthereumNftBridge = exports.MockNftBridge = void 0;
const anchor_1 = require("@project-serum/anchor");
const web3_js_1 = require("@solana/web3.js");
const utils_1 = require("../utils");
const wormhole_1 = require("./wormhole");
class MockNftBridge extends wormhole_1.MockEmitter {
    constructor(emitterAddress, chain, consistencyLevel) {
        super(emitterAddress, chain);
        this.consistencyLevel = consistencyLevel;
    }
    publishNftBridgeMessage(serialized, nonce, timestamp, uptickSequence = true) {
        return this.publishMessage(nonce == undefined ? 0 : nonce, serialized, this.consistencyLevel, timestamp, uptickSequence);
    }
    publishTransferNft(tokenAddress, tokenChain, name, symbol, tokenId, uri, recipientChain, recipient, nonce, timestamp, uptickSequence = true) {
        if (uri.length > 200) {
            throw new Error("uri.length > 200");
        }
        const serialized = Buffer.alloc(166 + uri.length);
        serialized.writeUInt8(1, 0);
        serialized.write(tokenAddress, 1, "hex");
        serialized.writeUInt16BE(tokenChain, 33);
        // truncate to 32 characters
        symbol = symbol.substring(0, 32);
        serialized.write(symbol, 35);
        // truncate to 32 characters
        name = name.substring(0, 32);
        serialized.write(name, 67);
        const tokenIdBytes = new anchor_1.BN(tokenId.toString()).toBuffer();
        serialized.write(tokenIdBytes.toString("hex"), 131 - tokenIdBytes.length, "hex");
        serialized.writeUInt8(uri.length, 131);
        serialized.write(uri, 132);
        const uriEnd = 132 + uri.length;
        serialized.write(recipient, uriEnd, "hex");
        serialized.writeUInt16BE(recipientChain, uriEnd + 32);
        return this.publishNftBridgeMessage(serialized, nonce, timestamp, uptickSequence);
    }
}
exports.MockNftBridge = MockNftBridge;
class MockEthereumNftBridge extends MockNftBridge {
    constructor(emitterAddress) {
        const chain = 2;
        super(utils_1.tryNativeToHexString(emitterAddress, chain), chain, 15);
    }
}
exports.MockEthereumNftBridge = MockEthereumNftBridge;
class MockSolanaNftBridge extends MockNftBridge {
    constructor(emitterAddress) {
        super(new web3_js_1.PublicKey(emitterAddress).toBuffer().toString("hex"), 1, 32);
    }
}
exports.MockSolanaNftBridge = MockSolanaNftBridge;
