// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcWasmApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcWasmApi = void 0;
const transformers_1 = require("../transformers");
const pagination_1 = require("../../../utils/pagination");
const types_1 = require("../types");
const exceptions_1 = require("@injectivelabs/exceptions");
const BaseGrpcWebConsumer_1 = require("../../BaseGrpcWebConsumer");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
/**
 * @category Chain Grpc API
 */
class ChainGrpcWasmApi {
    constructor(endpoint) {
        this.module = types_1.ChainModule.Wasm;
        this.client = new core_proto_ts_1.CosmwasmWasmV1Query.QueryClientImpl((0, BaseGrpcWebConsumer_1.getGrpcWebImpl)(endpoint));
    }
    fetchContractAccountsBalance({ contractAddress, pagination, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryAllContractStateRequest.create();
            request.address = contractAddress;
            const paginationForRequest = (0, pagination_1.paginationRequestFromPagination)(pagination);
            if (paginationForRequest) {
                request.pagination = paginationForRequest;
            }
            try {
                const response = yield this.client.AllContractState(request);
                return transformers_1.ChainGrpcWasmTransformer.allContractStateResponseToContractAccountsBalanceWithPagination(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchContractInfo(contractAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryContractInfoRequest.create();
            request.address = contractAddress;
            try {
                const response = yield this.client.ContractInfo(request);
                const contractInfo = response.contractInfo;
                if (!contractInfo) {
                    return;
                }
                return transformers_1.ChainGrpcWasmTransformer.contactInfoResponseToContractInfo(contractInfo);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchContractHistory(contractAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryContractHistoryRequest.create();
            request.address = contractAddress;
            try {
                const response = yield this.client.ContractHistory(request);
                return transformers_1.ChainGrpcWasmTransformer.contactHistoryResponseToContractHistory(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchSmartContractState(contractAddress, query) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QuerySmartContractStateRequest.create();
            request.address = contractAddress;
            if (query) {
                request.queryData = Buffer.from(query, 'base64');
            }
            try {
                const response = yield this.client.SmartContractState(request);
                return response;
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchRawContractState(contractAddress, query) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryRawContractStateRequest.create();
            request.address = contractAddress;
            if (query) {
                request.queryData = Buffer.from(query, 'base64');
            }
            try {
                const response = yield this.client.RawContractState(request);
                return response;
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchContractCodes(pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryCodesRequest.create();
            const paginationForRequest = (0, pagination_1.paginationRequestFromPagination)(pagination);
            if (paginationForRequest) {
                request.pagination = paginationForRequest;
            }
            try {
                const response = yield this.client.Codes(request);
                return transformers_1.ChainGrpcWasmTransformer.contractCodesResponseToContractCodes(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchContractCode(codeId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryCodeRequest.create();
            request.codeId = codeId.toString();
            try {
                const response = yield this.client.Code(request);
                return transformers_1.ChainGrpcWasmTransformer.contractCodeResponseToContractCode(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchContractCodeContracts(codeId, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmwasmWasmV1Query.QueryContractsByCodeRequest.create();
            request.codeId = codeId.toString();
            const paginationForRequest = (0, pagination_1.paginationRequestFromPagination)(pagination);
            if (paginationForRequest) {
                request.pagination = paginationForRequest;
            }
            try {
                const response = yield this.client.ContractsByCode(request);
                return transformers_1.ChainGrpcWasmTransformer.contractByCodeResponseToContractByCode(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmwasmWasmV1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
}
exports.ChainGrpcWasmApi = ChainGrpcWasmApi;
