// @injectivelabs/sdk-ts/dist/cjs/client/chain/grpc/ChainGrpcDistributionApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainGrpcDistributionApi = void 0;
const transformers_1 = require("../transformers");
const types_1 = require("../types");
const exceptions_1 = require("@injectivelabs/exceptions");
const BaseGrpcWebConsumer_1 = require("../../BaseGrpcWebConsumer");
const core_proto_ts_1 = require("@injectivelabs/core-proto-ts");
/**
 * @category Chain Grpc API
 */
class ChainGrpcDistributionApi {
    constructor(endpoint) {
        this.module = types_1.ChainModule.Distribution;
        this.client = new core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryClientImpl((0, BaseGrpcWebConsumer_1.getGrpcWebImpl)(endpoint));
    }
    fetchModuleParams() {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryParamsRequest.create();
            try {
                const response = yield this.client.Params(request);
                return transformers_1.ChainGrpcDistributionTransformer.moduleParamsResponseToModuleParams(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmosDistributionV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e, {
                    code: exceptions_1.UnspecifiedErrorCode,
                    contextModule: this.module,
                });
            }
        });
    }
    fetchDelegatorRewardsForValidator({ delegatorAddress, validatorAddress, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryDelegationRewardsRequest.create();
            request.validatorAddress = validatorAddress;
            request.delegatorAddress = delegatorAddress;
            try {
                const response = yield this.client.DelegationRewards(request);
                return transformers_1.ChainGrpcDistributionTransformer.delegationRewardResponseToReward(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmosDistributionV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e);
            }
        });
    }
    fetchDelegatorRewardsForValidatorNoThrow({ delegatorAddress, validatorAddress, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryDelegationRewardsRequest.create();
            request.validatorAddress = validatorAddress;
            request.delegatorAddress = delegatorAddress;
            try {
                const response = yield this.client.DelegationRewards(request);
                return transformers_1.ChainGrpcDistributionTransformer.delegationRewardResponseToReward(response);
            }
            catch (e) {
                if (e.message.includes('does not exist')) {
                    return [];
                }
                if (e instanceof core_proto_ts_1.CosmosDistributionV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e);
            }
        });
    }
    fetchDelegatorRewards(injectiveAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryDelegationTotalRewardsRequest.create();
            request.delegatorAddress = injectiveAddress;
            try {
                const response = yield this.client.DelegationTotalRewards(request);
                return transformers_1.ChainGrpcDistributionTransformer.totalDelegationRewardResponseToTotalReward(response);
            }
            catch (e) {
                if (e instanceof core_proto_ts_1.CosmosDistributionV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e);
            }
        });
    }
    fetchDelegatorRewardsNoThrow(injectiveAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = core_proto_ts_1.CosmosDistributionV1Beta1Query.QueryDelegationTotalRewardsRequest.create();
            request.delegatorAddress = injectiveAddress;
            try {
                const response = yield this.client.DelegationTotalRewards(request);
                return transformers_1.ChainGrpcDistributionTransformer.totalDelegationRewardResponseToTotalReward(response);
            }
            catch (e) {
                if (e.message.includes('does not exist')) {
                    return [];
                }
                if (e instanceof core_proto_ts_1.CosmosDistributionV1Beta1Query.GrpcWebError) {
                    throw new exceptions_1.GrpcUnaryRequestException(new Error(e.toString()), {
                        code: e.code,
                        contextModule: this.module,
                    });
                }
                throw new exceptions_1.GrpcUnaryRequestException(e);
            }
        });
    }
}
exports.ChainGrpcDistributionApi = ChainGrpcDistributionApi;
