// @injectivelabs/token-metadata/dist/cjs/tokens/tokens/testnet-tokens.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.testnetTokens = exports.overrideCw20s = void 0;
const tokens_1 = __importDefault(require("./tokens"));
const types_1 = require("../../types");
const overrideCw20s = (source, cw20sList) => {
    if (!cw20sList) {
        return [source];
    }
    const overrideCw20ItemIndex = cw20sList.findIndex(({ symbol }) => symbol === source.symbol);
    if (overrideCw20ItemIndex === -1) {
        return [...cw20sList, source];
    }
    cw20sList[overrideCw20ItemIndex] = source;
    return cw20sList;
};
exports.overrideCw20s = overrideCw20s;
const testnetTokens = () => ({
    wBTC: Object.assign(Object.assign({}, tokens_1.default.wBTC), { tokenFactories: [
            {
                symbol: 'wBTC',
                creator: 'inj17vytdwqczqz72j65saukplrktd4gyfme5agf6c',
                decimals: 8,
            },
        ] }),
    ATOM: Object.assign(Object.assign({}, tokens_1.default.ATOM), { tokenFactories: [
            {
                symbol: 'ATOM',
                creator: 'inj17vytdwqczqz72j65saukplrktd4gyfme5agf6c',
                decimals: 8,
            },
        ] }),
    INJ: Object.assign(Object.assign({}, tokens_1.default.INJ), { erc20: Object.assign(Object.assign({}, tokens_1.default.INJ.erc20), { address: '0xAD1794307245443B3Cb55d88e79EEE4d8a548C03' }) }),
    USDT: Object.assign(Object.assign({}, tokens_1.default.USDT), { erc20: Object.assign(Object.assign({}, tokens_1.default.USDT.erc20), { address: '0x87aB3B4C8661e07D6372361211B96ed4Dc36B1B5' }) }),
    APE: Object.assign(Object.assign({}, tokens_1.default.APE), { erc20: Object.assign(Object.assign({}, tokens_1.default.APE.erc20), { address: '0x44C21afAaF20c270EBbF5914Cfc3b5022173FEB7' }) }),
    wETH: Object.assign(Object.assign({}, tokens_1.default.wETH), { erc20: Object.assign(Object.assign({}, tokens_1.default.wETH.erc20), { address: '0xdB309Bb079EB419C18fe7D568c61cD2FdB65D9aF' }), tokenFactories: [
            {
                symbol: 'wETH',
                creator: 'inj17vytdwqczqz72j65saukplrktd4gyfme5agf6c',
                decimals: 8,
            },
        ] }),
    ASTRO: Object.assign(Object.assign({}, tokens_1.default.ASTRO), { ibcs: [
            Object.assign(Object.assign({}, (tokens_1.default.ASTRO.ibcs ? tokens_1.default.ASTRO.ibcs[0] : {})), { hash: 'E8AC6B792CDE60AB208CA060CA010A3881F682A7307F624347AB71B6A0B0BF89', path: 'transfer/channel-13', channelId: 'channel-13', baseDenom: 'ASTRO' }),
        ] }),
    SOL: Object.assign(Object.assign({}, tokens_1.default.SOL), { cw20s: [
            Object.assign(Object.assign({}, (tokens_1.default.SOL.cw20s ? tokens_1.default.SOL.cw20s[0] : {})), { address: 'inj12ngevx045zpvacus9s6anr258gkwpmthnz80e9' }),
        ] }),
    USDC: Object.assign(Object.assign({}, tokens_1.default.USDC), { erc20: Object.assign(Object.assign({}, tokens_1.default.USDC.erc20), { address: '0x07865c6E87B9F70255377e024ace6630C1Eaa37F' }), tokenFactories: [
            {
                symbol: 'USDC',
                creator: 'inj17vytdwqczqz72j65saukplrktd4gyfme5agf6c',
                decimals: 6,
            },
        ], cw20s: [
            ...(0, exports.overrideCw20s)({
                decimals: 6,
                symbol: 'USDCet',
                source: types_1.TokenSource.EthereumWh,
                address: 'inj12sqy9uzzl3h3vqxam7sz9f0yvmhampcgesh3qw',
            }, tokens_1.default.USDC.cw20s),
        ] }),
    MATIC: Object.assign(Object.assign({}, tokens_1.default.MATIC), { erc20: Object.assign(Object.assign({}, tokens_1.default.MATIC.erc20), { address: '0x9c3c9283d3e44854697cd22d3faa240cfb032889' }), tokenFactories: [
            {
                symbol: 'MATIC',
                creator: 'inj17vytdwqczqz72j65saukplrktd4gyfme5agf6c',
                decimals: 18,
            },
        ] }),
    STINJ: Object.assign(Object.assign({}, tokens_1.default.STINJ), { tokenFactories: [
            {
                symbol: 'STINJ',
                creator: 'inj17gkuet8f6pssxd8nycm3qr9d9y699rupv6397z',
                decimals: 18,
            },
        ] }),
    hINJ: Object.assign(Object.assign({}, tokens_1.default.hINJ), { cw20s: [
            ...(tokens_1.default.hINJ.cw20s || []),
            {
                symbol: 'hINJ',
                decimals: 18,
                address: 'inj1mz7mfhgx8tuvjqut03qdujrkzwlx9xhcj6yldc',
            },
        ] }),
    HDRO: Object.assign(Object.assign({}, tokens_1.default.HDRO), { tokenFactories: [
            ...(tokens_1.default.HDRO.tokenFactories || []),
            {
                creator: 'inj1pk7jhvjj2lufcghmvr7gl49dzwkk3xj0uqkwfk',
                symbol: 'hdro',
                decimals: 6,
            },
        ] }),
});
exports.testnetTokens = testnetTokens;
