// @certusone/wormhole-sdk/lib/cjs/solana/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWormholeCpiAccounts = exports.createBridgeFeeTransferInstruction = exports.createPostVaaInstructionSolana = exports.createVerifySignaturesInstructionsSolana = exports.postVaaSolanaWithRetry = exports.postVaaSolana = void 0;
__exportStar(require("./utils"), exports);
var sendAndConfirmPostVaa_1 = require("./sendAndConfirmPostVaa");
Object.defineProperty(exports, "postVaaSolana", { enumerable: true, get: function () { return sendAndConfirmPostVaa_1.postVaa; } });
Object.defineProperty(exports, "postVaaSolanaWithRetry", { enumerable: true, get: function () { return sendAndConfirmPostVaa_1.postVaaWithRetry; } });
var wormhole_1 = require("./wormhole");
Object.defineProperty(exports, "createVerifySignaturesInstructionsSolana", { enumerable: true, get: function () { return wormhole_1.createVerifySignaturesInstructions; } });
Object.defineProperty(exports, "createPostVaaInstructionSolana", { enumerable: true, get: function () { return wormhole_1.createPostVaaInstruction; } });
Object.defineProperty(exports, "createBridgeFeeTransferInstruction", { enumerable: true, get: function () { return wormhole_1.createBridgeFeeTransferInstruction; } });
Object.defineProperty(exports, "getWormholeCpiAccounts", { enumerable: true, get: function () { return wormhole_1.getPostMessageAccounts; } });
__exportStar(require("./wormhole/cpi"), exports);
__exportStar(require("./tokenBridge/cpi"), exports);
