// @across-protocol/sdk-v2/dist/cjs/priceClient/adapters/baseAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseHTTPAdapter = void 0;
var tslib_1 = require("tslib");
var assert_1 = tslib_1.__importDefault(require("assert"));
var axios_1 = tslib_1.__importDefault(require("axios"));
var BaseHTTPAdapter = (function () {
    function BaseHTTPAdapter(name, host, _a) {
        var _b = _a.timeout, timeout = _b === void 0 ? 1000 : _b, _c = _a.retries, retries = _c === void 0 ? 1 : _c;
        this.name = name;
        this.host = host;
        this._retries = 0;
        this._timeout = 0;
        this.retries = retries;
        this.timeout = timeout;
    }
    Object.defineProperty(BaseHTTPAdapter.prototype, "retries", {
        get: function () {
            return this._retries;
        },
        set: function (retries) {
            (0, assert_1.default)(retries >= 0);
            this._retries = retries;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(BaseHTTPAdapter.prototype, "timeout", {
        get: function () {
            return this._timeout;
        },
        set: function (timeout) {
            (0, assert_1.default)(timeout >= 0);
            this._timeout = timeout;
        },
        enumerable: false,
        configurable: true
    });
    BaseHTTPAdapter.prototype.query = function (path, urlArgs) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var url, args, errs, tries, err_1, errMsg;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = "https://".concat(this.host, "/").concat(path !== null && path !== void 0 ? path : "");
                        args = {
                            timeout: this.timeout,
                            params: urlArgs !== null && urlArgs !== void 0 ? urlArgs : {},
                        };
                        errs = [];
                        tries = 0;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 6]);
                        return [4, (0, axios_1.default)(url, args)];
                    case 2: return [2, (_a.sent()).data];
                    case 3:
                        err_1 = _a.sent();
                        errMsg = axios_1.default.isAxiosError(err_1) || err_1 instanceof Error ? err_1.message : "unknown error";
                        errs.push(errMsg);
                        if (!(++tries <= this.retries)) return [3, 5];
                        return [4, this.sleep(Math.pow(1.5, tries) * 1000)];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [3, 6];
                    case 6:
                        if (tries <= this.retries) return [3, 1];
                        _a.label = 7;
                    case 7: throw new Error("".concat(this.name, " price lookup failure (").concat(errs.join(", "), ")"));
                }
            });
        });
    };
    BaseHTTPAdapter.prototype.sleep = function (ms) {
        return new Promise(function (r) { return setTimeout(r, ms); });
    };
    return BaseHTTPAdapter;
}());
exports.BaseHTTPAdapter = BaseHTTPAdapter;
