// @wormhole-foundation/connect-sdk/dist/cjs/routes/route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isManual = exports.ManualRoute = exports.isAutomatic = exports.AutomaticRoute = exports.Route = void 0;
class Route {
    wh;
    request;
    constructor(wh, request) {
        this.wh = wh;
        this.request = request;
    }
    // Get the url to view the transfer on the explorer for the route provider
    transferUrl(txid) {
        return `https://wormholescan.io/#/tx/${txid}?network=${this.wh.network}`;
    }
}
exports.Route = Route;
class AutomaticRoute extends Route {
    IS_AUTOMATIC = true;
}
exports.AutomaticRoute = AutomaticRoute;
function isAutomatic(route) {
    return route.isAvailable !== undefined && route.IS_AUTOMATIC;
}
exports.isAutomatic = isAutomatic;
class ManualRoute extends Route {
    NATIVE_GAS_DROPOFF_SUPPORTED = false;
    IS_AUTOMATIC = false;
}
exports.ManualRoute = ManualRoute;
function isManual(route) {
    return route.complete !== undefined;
}
exports.isManual = isManual;
