// @injectivelabs/sdk-ts/dist/cjs/client/indexer/types/explorer-rest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatorUptimeStatus = exports.AccessType = exports.AccessTypeCode = void 0;
var AccessTypeCode;
(function (AccessTypeCode) {
    AccessTypeCode[AccessTypeCode["AccessTypeUnspecified"] = 0] = "AccessTypeUnspecified";
    AccessTypeCode[AccessTypeCode["AccessTypeNobody"] = 1] = "AccessTypeNobody";
    AccessTypeCode[AccessTypeCode["AccessTypeOnlyAddress"] = 2] = "AccessTypeOnlyAddress";
    AccessTypeCode[AccessTypeCode["AccessTypeEverybody"] = 3] = "AccessTypeEverybody";
})(AccessTypeCode = exports.AccessTypeCode || (exports.AccessTypeCode = {}));
var AccessType;
(function (AccessType) {
    AccessType["AccessTypeUnspecified"] = "Unspecified";
    AccessType["AccessTypeNobody"] = "Nobody";
    AccessType["AccessTypeOnlyAddress"] = "Only Address";
    AccessType["AccessTypeEverybody"] = "Everybody";
})(AccessType = exports.AccessType || (exports.AccessType = {}));
var ValidatorUptimeStatus;
(function (ValidatorUptimeStatus) {
    ValidatorUptimeStatus["Proposed"] = "proposed";
    ValidatorUptimeStatus["Signed"] = "signed";
    ValidatorUptimeStatus["Missed"] = "missed";
})(ValidatorUptimeStatus = exports.ValidatorUptimeStatus || (exports.ValidatorUptimeStatus = {}));
