// @injectivelabs/sdk-ts/dist/cjs/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_STD_FEE_BY_DENOM = exports.DEFAULT_STD_FEE = exports.CW20_ADAPTER_CONTRACT_BY_NETWORK = exports.DEFAULT_PAGINATION_TOTAL_COUNT = exports.PAGINATION_TOTAL_PAGE_LIMIT = exports.DUST_AMOUNT = exports.DEFAULT_DERIVATION_PATH = exports.BECH32_ADDR_CONS_PREFIX = exports.BECH32_ADDR_VAL_PREFIX = exports.BECH32_ADDR_ACC_PREFIX = exports.BECH32_PUBKEY_CONS_PREFIX = exports.BECH32_PUBKEY_VAL_PREFIX = exports.BECH32_PUBKEY_ACC_PREFIX = void 0;
const networks_1 = require("@injectivelabs/networks");
const utils_1 = require("@injectivelabs/utils");
Object.defineProperty(exports, "DEFAULT_STD_FEE", { enumerable: true, get: function () { return utils_1.DEFAULT_STD_FEE; } });
Object.defineProperty(exports, "DEFAULT_STD_FEE_BY_DENOM", { enumerable: true, get: function () { return utils_1.DEFAULT_STD_FEE_BY_DENOM; } });
exports.BECH32_PUBKEY_ACC_PREFIX = 'injpub';
exports.BECH32_PUBKEY_VAL_PREFIX = 'injvaloperpub';
exports.BECH32_PUBKEY_CONS_PREFIX = 'injvalconspub';
exports.BECH32_ADDR_ACC_PREFIX = 'inj';
exports.BECH32_ADDR_VAL_PREFIX = 'injvaloper';
exports.BECH32_ADDR_CONS_PREFIX = 'injvalcons';
exports.DEFAULT_DERIVATION_PATH = "m/44'/60'/0'/0/0";
exports.DUST_AMOUNT = 0.0001;
exports.PAGINATION_TOTAL_PAGE_LIMIT = 10000;
exports.DEFAULT_PAGINATION_TOTAL_COUNT = 1000000;
exports.CW20_ADAPTER_CONTRACT_BY_NETWORK = {
    [networks_1.Network.Mainnet]: 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk',
    [networks_1.Network.MainnetLB]: 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk',
    [networks_1.Network.MainnetK8s]: 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk',
    [networks_1.Network.Public]: 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk',
    [networks_1.Network.Staging]: 'inj14ejqjyq8um4p3xfqj74yld5waqljf88f9eneuk',
    [networks_1.Network.Testnet]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.TestnetK8s]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.TestnetOld]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.Devnet]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.Devnet1]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.Devnet2]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
    [networks_1.Network.Local]: 'inj1hdvy6tl89llqy3ze8lv6mz5qh66sx9enn0jxg6',
};
