// @certusone/wormhole-sdk/lib/cjs/solana/tokenBridge/accounts/endpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointRegistration = exports.getEndpointRegistration = exports.deriveEndpointKey = void 0;
const web3_js_1 = require("@solana/web3.js");
const utils_1 = require("../../../utils");
const utils_2 = require("../../utils");
function deriveEndpointKey(tokenBridgeProgramId, emitterChain, emitterAddress) {
    if (emitterChain == utils_1.CHAIN_ID_SOLANA) {
        throw new Error("emitterChain == CHAIN_ID_SOLANA cannot exist as foreign token bridge emitter");
    }
    if (typeof emitterAddress == "string") {
        emitterAddress = utils_1.tryNativeToUint8Array(emitterAddress, emitterChain);
    }
    return utils_2.deriveAddress([
        (() => {
            const buf = Buffer.alloc(2);
            buf.writeUInt16BE(emitterChain);
            return buf;
        })(),
        emitterAddress,
    ], tokenBridgeProgramId);
}
exports.deriveEndpointKey = deriveEndpointKey;
function getEndpointRegistration(connection, endpointKey, commitment) {
    return __awaiter(this, void 0, void 0, function* () {
        return connection
            .getAccountInfo(new web3_js_1.PublicKey(endpointKey), commitment)
            .then((info) => EndpointRegistration.deserialize(utils_2.getAccountData(info)));
    });
}
exports.getEndpointRegistration = getEndpointRegistration;
class EndpointRegistration {
    constructor(chain, contract) {
        this.chain = chain;
        this.contract = contract;
    }
    static deserialize(data) {
        if (data.length != 34) {
            throw new Error("data.length != 34");
        }
        const chain = data.readUInt16LE(0);
        const contract = data.subarray(2, 34);
        return new EndpointRegistration(chain, contract);
    }
}
exports.EndpointRegistration = EndpointRegistration;
