// @across-protocol/sdk-v2/dist/cjs/clients/mocks/MockConfigStoreClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockConfigStoreClient = void 0;
var tslib_1 = require("tslib");
var assert_1 = tslib_1.__importDefault(require("assert"));
var ethers_1 = require("ethers");
var utils_1 = require("../../utils");
var AcrossConfigStoreClient_1 = require("../AcrossConfigStoreClient");
var MockEvents_1 = require("./MockEvents");
var MockConfigStoreClient = (function (_super) {
    tslib_1.__extends(MockConfigStoreClient, _super);
    function MockConfigStoreClient(logger, configStore, eventSearchConfig, configStoreVersion, chainId, mockUpdate, availableChainIdsOverride) {
        if (eventSearchConfig === void 0) { eventSearchConfig = { fromBlock: 0, maxBlockLookBack: 0 }; }
        if (chainId === void 0) { chainId = 1; }
        if (mockUpdate === void 0) { mockUpdate = false; }
        var _this = _super.call(this, logger, configStore, eventSearchConfig, configStoreVersion) || this;
        _this.configStoreVersion = AcrossConfigStoreClient_1.DEFAULT_CONFIG_STORE_VERSION;
        _this.eventSignatures = {
            OwnershipTransferred: "address,address",
            UpdatedGlobalConfig: "bytes32,string",
            UpdatedTokenConfig: "address,string",
        };
        _this.chainId = chainId;
        _this.eventManager = mockUpdate ? (0, MockEvents_1.getEventManager)(chainId, _this.eventSignatures) : null;
        if ((0, utils_1.isDefined)(_this.eventManager) && _this.eventManager) {
            _this.updateGlobalConfig(AcrossConfigStoreClient_1.GLOBAL_CONFIG_STORE_KEYS.CHAIN_ID_INDICES, JSON.stringify(availableChainIdsOverride), {
                blockNumber: _this.eventManager.blockNumber,
            });
        }
        return _this;
    }
    MockConfigStoreClient.prototype.setAvailableChains = function (chainIds) {
        this.availableChainIdsOverride = chainIds;
    };
    MockConfigStoreClient.prototype.getChainIdIndicesForBlock = function (block) {
        var _a;
        return (_a = this.availableChainIdsOverride) !== null && _a !== void 0 ? _a : _super.prototype.getChainIdIndicesForBlock.call(this, block);
    };
    MockConfigStoreClient.prototype.getConfigStoreVersionForBlock = function (_blockNumber) {
        return this.configStoreVersion === AcrossConfigStoreClient_1.DEFAULT_CONFIG_STORE_VERSION
            ? _super.prototype.getConfigStoreVersionForBlock.call(this, _blockNumber)
            : this.configStoreVersion;
    };
    MockConfigStoreClient.prototype.setConfigStoreVersion = function (version) {
        this.configStoreVersion = version;
    };
    MockConfigStoreClient.prototype._update = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var eventNames, latestBlockSearched, globalConfigUpdateTimes, _events, _i, _a, event_1, idx, block, events;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.eventManager === null) {
                            return [2, _super.prototype._update.call(this)];
                        }
                        eventNames = ["UpdatedGlobalConfig", "UpdatedTokenConfig"];
                        latestBlockSearched = this.eventManager.blockNumber;
                        globalConfigUpdateTimes = [];
                        _events = eventNames.map(function () { return []; });
                        _i = 0, _a = this.eventManager.getEvents().flat();
                        _b.label = 1;
                    case 1:
                        if (!(_i < _a.length)) return [3, 4];
                        event_1 = _a[_i];
                        idx = eventNames.indexOf(event_1.event);
                        if (idx !== -1) {
                            _events[idx].push(event_1);
                        }
                        if (!(event_1.event === "UpdatedGlobalConfig")) return [3, 3];
                        return [4, event_1.getBlock()];
                    case 2:
                        block = _b.sent();
                        globalConfigUpdateTimes.push(block.timestamp);
                        _b.label = 3;
                    case 3:
                        _i++;
                        return [3, 1];
                    case 4:
                        events = Object.fromEntries(eventNames.map(function (eventName, idx) { return [eventName, _events[idx]]; }));
                        return [2, {
                                success: true,
                                chainId: this.chainId,
                                searchEndBlock: this.eventSearchConfig.toBlock || latestBlockSearched,
                                events: {
                                    updatedGlobalConfigEvents: events["UpdatedGlobalConfig"],
                                    globalConfigUpdateTimes: globalConfigUpdateTimes,
                                    updatedTokenConfigEvents: events["UpdatedTokenConfig"],
                                },
                            }];
                }
            });
        });
    };
    MockConfigStoreClient.prototype.updateGlobalConfig = function (key, value, overrides) {
        if (overrides === void 0) { overrides = {}; }
        return this.generateConfig("UpdatedGlobalConfig", (0, utils_1.utf8ToHex)(key), value, overrides);
    };
    MockConfigStoreClient.prototype.updateTokenConfig = function (key, value, overrides) {
        if (overrides === void 0) { overrides = {}; }
        if (ethers_1.ethers.utils.isAddress(key) === false) {
            throw new Error("Invalid address: ".concat(key));
        }
        return this.generateConfig("UpdatedTokenConfig", key, value, overrides);
    };
    MockConfigStoreClient.prototype.generateConfig = function (event, key, value, overrides) {
        if (overrides === void 0) { overrides = {}; }
        (0, assert_1.default)(this.eventManager !== null);
        var topics = [key, value];
        var args = { key: key, value: value };
        return this.eventManager.generateEvent({
            event: event,
            address: this.configStore.address,
            topics: topics.map(function (topic) { return topic.toString(); }),
            args: args,
            blockNumber: overrides.blockNumber,
        });
    };
    return MockConfigStoreClient;
}(AcrossConfigStoreClient_1.AcrossConfigStoreClient));
exports.MockConfigStoreClient = MockConfigStoreClient;
